<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Patrick Cherry">
<meta name="dcterms.date" content="2024-01-22">

<title>Frameshift – scRNA-seq Analysis: Messy Mouse Brain</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script src="../site_libs/quarto-contrib/iconify-2.1.0/iconify-icon.min.js"></script>
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<meta name="fediverse:creator" content="@pcherry@fosstodon.org">


<link rel="stylesheet" href="../styles.css">
<meta property="og:title" content="Frameshift - scRNA-seq Analysis: Messy Mouse Brain">
<meta property="og:description" content="">
<meta property="og:site_name" content="Frameshift">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Frameshift</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Patrick-Cherry-cv.html"> 
<span class="menu-text">CV</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/pdcherry"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://fosstodon.org/@pcherry"> <i class="bi bi-mastodon" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://pixelfed.social/pcherry"> <i class="bi bi-instagram" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/p-cherry/"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://bsky.app/profile/pcherry.bsky.social"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="fa6-brands:bluesky" aria-label="Icon bluesky from fa6-brands Iconify.design set." title="Icon bluesky from fa6-brands Iconify.design set."></iconify-icon></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">scRNA-seq Analysis: Messy Mouse Brain</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">code</div>
                <div class="quarto-category">scRNA-seq</div>
                <div class="quarto-category">RNA-seq</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Patrick Cherry </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">January 22, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul class="collapse">
  <li><a href="#objectives" id="toc-objectives" class="nav-link" data-scroll-target="#objectives">Objectives</a></li>
  </ul></li>
  <li><a href="#data-description" id="toc-data-description" class="nav-link" data-scroll-target="#data-description">Data description</a></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  <li><a href="#data-qc" id="toc-data-qc" class="nav-link" data-scroll-target="#data-qc">Data QC</a>
  <ul class="collapse">
  <li><a href="#features" id="toc-features" class="nav-link" data-scroll-target="#features">Features</a></li>
  <li><a href="#filter-on-feature-count" id="toc-filter-on-feature-count" class="nav-link" data-scroll-target="#filter-on-feature-count">Filter on feature count</a>
  <ul class="collapse">
  <li><a href="#features-1" id="toc-features-1" class="nav-link" data-scroll-target="#features-1">Features</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#initial-analysis" id="toc-initial-analysis" class="nav-link" data-scroll-target="#initial-analysis">Initial Analysis</a>
  <ul class="collapse">
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization">Normalization</a></li>
  <li><a href="#feature-selection-by-pca" id="toc-feature-selection-by-pca" class="nav-link" data-scroll-target="#feature-selection-by-pca">Feature Selection by PCA</a></li>
  <li><a href="#scale" id="toc-scale" class="nav-link" data-scroll-target="#scale">Scale</a></li>
  <li><a href="#run-pca" id="toc-run-pca" class="nav-link" data-scroll-target="#run-pca">Run PCA</a>
  <ul class="collapse">
  <li><a href="#visualize-pca-results" id="toc-visualize-pca-results" class="nav-link" data-scroll-target="#visualize-pca-results">Visualize PCA Results</a></li>
  </ul></li>
  <li><a href="#determine-the-dimensionality-of-the-dataset" id="toc-determine-the-dimensionality-of-the-dataset" class="nav-link" data-scroll-target="#determine-the-dimensionality-of-the-dataset">Determine the dimensionality of the dataset</a></li>
  </ul></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering">Clustering</a></li>
  <li><a href="#run-non-linear-dimensional-reduction-umaptsne" id="toc-run-non-linear-dimensional-reduction-umaptsne" class="nav-link" data-scroll-target="#run-non-linear-dimensional-reduction-umaptsne">Run non-linear dimensional reduction (UMAP/tSNE)</a>
  <ul class="collapse">
  <li><a href="#identify-cluster-biomarkers" id="toc-identify-cluster-biomarkers" class="nav-link" data-scroll-target="#identify-cluster-biomarkers">Identify cluster biomarkers</a>
  <ul class="collapse">
  <li><a href="#cluster-3" id="toc-cluster-3" class="nav-link" data-scroll-target="#cluster-3">Cluster 3</a></li>
  <li><a href="#all-marker-analysis" id="toc-all-marker-analysis" class="nav-link" data-scroll-target="#all-marker-analysis">All marker analysis</a></li>
  <li><a href="#set-up-function-for-all" id="toc-set-up-function-for-all" class="nav-link" data-scroll-target="#set-up-function-for-all">Set up function for all</a></li>
  <li><a href="#run-the-functon-in-df-env" id="toc-run-the-functon-in-df-env" class="nav-link" data-scroll-target="#run-the-functon-in-df-env">Run the functon in df env</a></li>
  <li><a href="#all-marker-analysis-1" id="toc-all-marker-analysis-1" class="nav-link" data-scroll-target="#all-marker-analysis-1">All marker analysis</a></li>
  <li><a href="#run-the-functon-in-df-env-1" id="toc-run-the-functon-in-df-env-1" class="nav-link" data-scroll-target="#run-the-functon-in-df-env-1">Run the functon in df env</a></li>
  <li><a href="#cluster-7" id="toc-cluster-7" class="nav-link" data-scroll-target="#cluster-7">Cluster 7</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#cell-type-assignments" id="toc-cell-type-assignments" class="nav-link" data-scroll-target="#cell-type-assignments">Cell type assignments</a>
  <ul class="collapse">
  <li><a href="#using-garnett" id="toc-using-garnett" class="nav-link" data-scroll-target="#using-garnett">Using garnett</a>
  <ul class="collapse">
  <li><a href="#classifier" id="toc-classifier" class="nav-link" data-scroll-target="#classifier">Classifier</a></li>
  <li><a href="#prepare-scrna-seq-data-for-garnett" id="toc-prepare-scrna-seq-data-for-garnett" class="nav-link" data-scroll-target="#prepare-scrna-seq-data-for-garnett">Prepare scRNA-seq data for garnett</a></li>
  <li><a href="#prepare-trained-classified-from-heirarchical-marker-gene-file" id="toc-prepare-trained-classified-from-heirarchical-marker-gene-file" class="nav-link" data-scroll-target="#prepare-trained-classified-from-heirarchical-marker-gene-file">Prepare trained classified from heirarchical marker gene file</a></li>
  <li><a href="#prepare-trained-classified-from-heirarchical-marker-gene-file-1" id="toc-prepare-trained-classified-from-heirarchical-marker-gene-file-1" class="nav-link" data-scroll-target="#prepare-trained-classified-from-heirarchical-marker-gene-file-1">Prepare trained classified from heirarchical marker gene file</a></li>
  <li><a href="#train-the-classifier" id="toc-train-the-classifier" class="nav-link" data-scroll-target="#train-the-classifier">Train the classifier</a></li>
  <li><a href="#perform-garnett-classify" id="toc-perform-garnett-classify" class="nav-link" data-scroll-target="#perform-garnett-classify">Perform garnett classify</a></li>
  </ul></li>
  <li><a href="#htp-go-anlaysis-for-all-10-clusters" id="toc-htp-go-anlaysis-for-all-10-clusters" class="nav-link" data-scroll-target="#htp-go-anlaysis-for-all-10-clusters">HTP GO Anlaysis for all 10 clusters</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>In this data analysis exercise, I analyze a mysterious 10x single-cell RNA-seq data file, which consists of a collection of single cells from an unknown tissue with unknown cell types. This is standard 10x mouse single cell RNA-seq data output and we are curious what type of tissue and cells these are.</p>
<section id="objectives" class="level3">
<h3 class="anchored" data-anchor-id="objectives">Objectives</h3>
<ol type="1">
<li><strong>Quality control and filtering.</strong> Perform a basic QC of the data, explaining the rationale behind each step and the choice of parameters. How do the data look? What would you examine before moving forward to downstream analysis? Why?</li>
<li><strong>Dimension reduction and visualization.</strong> How can we make sense of all this data? Perform dimensional reduction with any methods at your choice, explaining the rationale behind the choice and caveats of other potential dimensional reduction methods. Plot these out in a suitable visualization(s).</li>
<li><strong>Marker genes, clusters tissue, and cell types.</strong> Identify genes and gene-sets that define the tissue and are differentially expressed. What tissue are we looking at? How many cell types are present, and what kind?</li>
</ol>
<p>Cell states can be very dynamic within the same cell type. For some of the most abundant cell types, feel free to dig deeper on any observed heterogeneity and the biological underpinning.</p>
</section>
</section>
<section id="data-description" class="level2">
<h2 class="anchored" data-anchor-id="data-description">Data description</h2>
<ul>
<li>counts: matrix <code>matrix.mtx.gz</code></li>
<li>cell barcodes: <code>barcodes.tsv.gz</code></li>
<li>gene features: <code>features.tsv.gz</code></li>
</ul>
</section>
<section id="load-data" class="level1">
<h1>Load data</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>koaladata <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="at">data.dir =</span> <span class="st">"../../scRNA-seq/Koala/"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>koala <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> koaladata, <span class="at">project =</span> <span class="st">"koala"</span>,</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.cells =</span> <span class="dv">3</span>, <span class="at">min.features =</span> <span class="dv">200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The following parameters are used: - min.cells = 3: a feature must have appeared in at least 3 cell barcodes - min.features = 200 a cell must have at least 200 features to be included</p>
</section>
<section id="data-qc" class="level1">
<h1>Data QC</h1>
<p>Here are some features in the metadata we can use for QC:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>koala<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">colnames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "orig.ident"   "nCount_RNA"   "nFeature_RNA"</code></pre>
</div>
</div>
<p>Unfortunately, a <em>“percent.mt”</em> vector is not available. I ordinarily would heavily QC on this information.</p>
<section id="features" class="level2">
<h2 class="anchored" data-anchor-id="features">Features</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(koala, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>), <span class="at">pt.size =</span> .<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">1100</span>) <span class="sc">+</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2000</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>koala<span class="sc">@</span>meta.data<span class="sc">$</span>nCount_RNA <span class="sc">%&gt;%</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">"koala"</span>, <span class="at">y =</span> value)) <span class="sc">+</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">fill =</span> <span class="st">"pink"</span>) <span class="sc">+</span> <span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="fl">0.2</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"nCount_RNA"</span>) <span class="sc">+</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>These are both skewed distributions, but RNA counts is more extremely skewed than features. Let’s see how they scatter:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>features_by_counts_scatter_table <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(koala<span class="sc">$</span>nCount_RNA, koala<span class="sc">$</span>nFeature_RNA) <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="st">"nCount_RNA"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"nFeature_RNA"</span> <span class="ot">=</span> <span class="dv">2</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>features_by_counts_scatter <span class="ot">&lt;-</span> <span class="fu">FeatureScatter</span>(</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  koala, <span class="at">feature1 =</span> <span class="st">"nCount_RNA"</span>, <span class="at">feature2 =</span> <span class="st">"nFeature_RNA"</span>) <span class="co">#+</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># stat_smooth(formula = "y ~ x*log(x)", method = glm) +</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># stat_smooth(formula = "y ~ log(x)", method = glm, color = "yellow") +</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># geom_smooth(formula = y ~ (Vm * x)/(K + x),</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>  <span class="co">#             #data = features_by_counts_scatter_table,</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>  <span class="co">#             method = "nls", color = "green", se = FALSE,</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#             method.args = list(start = c(K = 10000, Vm = 6000)))</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>features_by_counts_scatter</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<!-- ```{r} -->
<!-- summary(nls(formula = "nFeature_RNA ~ (Vm * nCount_RNA)/(K + nCount_RNA)", -->
<!--             data = features_by_counts_scatter_table, -->
<!--             start = c(K = 10000, Vm = 6000))) -->
<!-- ``` -->
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>features_by_counts_scatter <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The feature count appears to increase approximately logarithmicly with RNA count. The Pearson correlation is 0.95, but the non linearity indicates the relationship may be stronger than the correlation coefficient indicates. This observation is consistent with an RNA-seq experiment wherein there is a fixed number of potentially expressed RNAs and the number of detected transcripts begins to saturate the and approach the maximum number able to be detected.</p>
<p>The log-log transformed axis plot shows: - There are only a few cells/droplets with fewer than 300 RNA counts, which is great for an scRNA-seq experiment. - There is still some curve to the line, meaning the relationship is more trancendental (not simply a power-law, like polynomial).</p>
</section>
<section id="filter-on-feature-count" class="level2">
<h2 class="anchored" data-anchor-id="filter-on-feature-count">Filter on feature count</h2>
<p>I am going to keep droplets (“cells”) with 1100 features or fewer</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>koala_filtered <span class="ot">&lt;-</span> <span class="fu">subset</span>(koala, nFeature_RNA <span class="sc">&lt;=</span> <span class="dv">1100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s quickly confirm that the filtering changed the number of features throughout the Seurat object by using Seurat size commands to query the number of observations in the pre-filter and post-filter Seurat objects:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(koala<span class="sc">$</span>orig.ident); <span class="fu">length</span>(koala_filtered<span class="sc">$</span>orig.ident)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4449</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1410</code></pre>
</div>
</div>
<p>The feature count decreased from 4449 to 1410 upon filtering.</p>
<p>(The <code>orig.ident</code> matrix within the Seurat object stores the cell metadata, and counting its number of entries can tell us number of unique observations.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(koala_filtered<span class="sc">$</span>orig.ident)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ATGGCTGTCGATTTGA-1 GGTACAAAGTTCCTGC-1 ATTCAACCATAGGCGA-1 TCTCCTCGTCAATACG-1 
             koala              koala              koala              koala 
TTTCTCACAGCACCAT-1 TTCCTTGAGGATCACT-1 
             koala              koala 
Levels: koala</code></pre>
</div>
</div>
<section id="features-1" class="level3">
<h3 class="anchored" data-anchor-id="features-1">Features</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(koala_filtered, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>), <span class="at">pt.size =</span> .<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Post filtering, the RNA count plot looks more typical, with one major “peak” and a long tail. The features plot looks fairly evenly distributed because of the filtering cut off so close to the “peak” of the distribution, but I hypothesize it was necessary due to the hypothesized doublet peak right at 2000 features per droplet right near by.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>features_by_counts_scatter_table <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(koala_filtered<span class="sc">$</span>nCount_RNA,</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>                                              koala_filtered<span class="sc">$</span>nFeature_RNA) <span class="sc">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="st">"nCount_RNA"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"nFeature_RNA"</span> <span class="ot">=</span> <span class="dv">2</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>features_by_counts_scatter <span class="ot">&lt;-</span> <span class="fu">FeatureScatter</span>(</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  koala_filtered, <span class="at">feature1 =</span> <span class="st">"nCount_RNA"</span>, <span class="at">feature2 =</span> <span class="st">"nFeature_RNA"</span>) <span class="co">#+</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># stat_smooth(formula = "y ~ x*log(x)", method = glm) +</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># stat_smooth(formula = "y ~ log(x)", method = glm, color = "yellow") +</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># geom_smooth(formula = y ~ (Vm * x)/(K + x),</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>  <span class="co">#             #data = features_by_counts_scatter_table,</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#             method = "nls", color = "green", se = FALSE,</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>  <span class="co">#             method.args = list(start = c(K = 10000, Vm = 6000)))</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>features_by_counts_scatter</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="initial-analysis" class="level1">
<h1>Initial Analysis</h1>
<section id="normalization" class="level2">
<h2 class="anchored" data-anchor-id="normalization">Normalization</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>koala_norm <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(koala_filtered)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts</code></pre>
</div>
</div>
<p>The above is a log-normalization. This procedure divides each feature’s abundance by the sample mean for that cell and takes the natural log (<code>log()</code> in <code>R</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>koala_features <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(koala_norm, <span class="at">selection.method =</span> <span class="st">"vst"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts</code></pre>
</div>
</div>
<p>The vst method is the default feature selection method, but I’m just making it explicit, because a later (cluster identification) package refers to this fact. The variance stabilizing transformation makes the variances (standard deviations) more uniform and stops them from being related to the mean value within a group, which facilitates downstream linear regression-based techniques (<em>e.g.</em> PCA).</p>
</section>
<section id="feature-selection-by-pca" class="level2">
<h2 class="anchored" data-anchor-id="feature-selection-by-pca">Feature Selection by PCA</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>top_koala_features <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">VariableFeatures</span>(koala_features), <span class="dv">10</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>top_koala_features_plot <span class="ot">&lt;-</span> <span class="fu">VariableFeaturePlot</span>(koala_features) <span class="sc">%&gt;%</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">LabelPoints</span>(<span class="at">points =</span> top_koala_features, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>top_koala_features_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Transformation introduced infinite values in continuous x-axis</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>top_koala_features</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Nrxn1"  "Ank3"   "Nkain2" "Sntg1"  "Glul"   "Rspo3"  "Reln"   "Aoah"  
 [9] "Slc1a2" "Spp1"  </code></pre>
</div>
</div>
<p>The most highly expressed gene is MEG3 (maternally expressed 3), a maternally expressed, imprinted long non-coding RNA (lncRNA). This likely means the tissue analyzed here is an egg cell, recently fertilized egg, or young embryo.</p>
</section>
<section id="scale" class="level2">
<h2 class="anchored" data-anchor-id="scale">Scale</h2>
<p>Pre-process to center at 0 and make sd = 1 prior to PCA.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>gene_names <span class="ot">&lt;-</span> <span class="fu">rownames</span>(koala_features)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>koala_features <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(koala_features, <span class="at">features =</span> gene_names)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
</div>
</section>
<section id="run-pca" class="level2">
<h2 class="anchored" data-anchor-id="run-pca">Run PCA</h2>
<p>Linear dimensionality reduction</p>
<p>I’m going to go with all features and use the Elbow plot to determine which features are still important after this step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>koala_pca <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(koala_features, <span class="at">features =</span> <span class="fu">VariableFeatures</span>(koala_features) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="visualize-pca-results" class="level3">
<h3 class="anchored" data-anchor-id="visualize-pca-results">Visualize PCA Results</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(koala_pca[[<span class="st">"pca"</span>]], <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>PC_ 1 
Positive:  Ptprb, Lrmda, Prkg1, Plekhg1, Cemip2, Plpp1, Rasal2, Meis2, Plxnc1, Adam23 
       Sptbn1, Stox2, Ets1, Mrc1, F8, Fyn, Syne1, Hecw2, Slc9a9, Fbxl7 
Negative:  Pck1, Ppara, Ttc39c, Gm29966, G6pc, Abcc2, Mlxipl, Errfi1, Egfr, Gfra1 
       Ahsg, Neat1, Trf, Slc27a2, Fam214a, Slc38a3, Slc7a2, Dpyd, Hc, Alas1 
PC_ 2 
Positive:  Nrxn1, Reln, Rbms3, Dcn, Ntm, Bmp5, Lhfp, Abcc9, Arhgap24, Nkain2 
       Pde3a, Tln2, Ngf, Sox5, Pde1a, Robo2, Hand2os1, Hgf, Ccbe1, Agtr1a 
Negative:  Ptprb, Rasal2, Cemip2, Plpp1, Adam23, Stox2, Fyn, Mrc1, Kdr, F8 
       Prickle1, Dab2, Adgrf5, Cyyr1, Plcb4, Hecw2, Maf, Arhgap31, Nrp1, Cd55 
PC_ 3 
Positive:  Slc8a1, Elmo1, Ptprc, Runx1, Cd5l, Clec4f, Dock2, Adgb, Entpd1, Dock10 
       Aoah, Gm26740, Bank1, Adgre1, Arhgap15, Ikzf1, E230029C05Rik, Epsti1, Adgre4, Cd163 
Negative:  Ptprb, Plekhg1, Cemip2, Meis2, Tshz2, Prkg1, Pard3b, Plpp1, Adgrl2, Plxnc1 
       Stox2, Rasal2, F8, Prickle1, Kdr, Hecw2, Fbxl7, Sptbn1, Adam23, Syne1 
PC_ 4 
Positive:  Gm2682, Skap1, Themis, Cd226, Satb1, Kcnq5, Ms4a4b, Grap2, Dock2, Arhgap15 
       Bcl11b, Ripor2, Inpp4b, Ipcef1, Il7r, Atp8a2, Lef1, Camk4, Cyfip2, Ikzf3 
Negative:  Clec4f, Slc8a1, Frmd4b, Aoah, Slc40a1, Cd5l, Tcf7l2, Kcnk13, Tbxas1, Adgb 
       Myo9a, Hdac9, Entpd1, Pde7b, Cd86, Adgre1, Lgmn, Ksr2, Zeb2, E230029C05Rik </code></pre>
</div>
</div>
<p>Quick print out all of the positive and negative features in the top four fitted principle components.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>(dim_reduc_plot <span class="ot">&lt;-</span> <span class="fu">VizDimLoadings</span>(koala_pca, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">reduction =</span> <span class="st">"pca"</span>) <span class="sc">&amp;</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">5</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>(pc_1_2_scatterplot <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(koala_pca, <span class="at">reduction =</span> <span class="st">"pca"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimHeatmap</span>(koala_pca, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="determine-the-dimensionality-of-the-dataset" class="level2">
<h2 class="anchored" data-anchor-id="determine-the-dimensionality-of-the-dataset">Determine the dimensionality of the dataset</h2>
<p>Using jackstraw</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>koala_js <span class="ot">&lt;-</span> <span class="fu">JackStraw</span>(koala_pca, <span class="at">num.replicate =</span> <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>koala_js_score <span class="ot">&lt;-</span> <span class="fu">ScoreJackStraw</span>(koala_js, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">JackStrawPlot</span>(koala_js_score, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 28212 rows containing missing values (`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(koala_js_score)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Deciding that the elbow occurs at 9 principle components, so should retain dimensions 1 to 10. However, due to the strange behavior of this dataset, I am going to run <code>FindClusters()</code> on all 20 principle components and see what happens. It may confirm my findings without intervention.</p>
</section>
</section>
<section id="clustering" class="level1">
<h1>Clustering</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>koala_clusters <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(koala_pca, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>koala_clusters <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(koala_clusters, <span class="at">resolution =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 1410
Number of edges: 53206

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8554
Number of communities: 9
Elapsed time: 0 seconds</code></pre>
</div>
</div>
<p>Despite retaining 20 dimensions in PCA, the Louvain cluster determination algorithm only identified 9 independently-clustering communities of droplets/cells. This is consistent with my by-eye reading of the elbow plot.</p>
</section>
<section id="run-non-linear-dimensional-reduction-umaptsne" class="level1">
<h1>Run non-linear dimensional reduction (UMAP/tSNE)</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>koala_umap <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(koala_clusters, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>23:49:50 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>23:49:50 Read 1410 rows and found 10 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>23:49:50 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>23:49:50 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
23:49:50 Writing NN index file to temp file /var/folders/11/sfzn5hhx2995d534f55tvzd40000gn/T//RtmpGrv29G/file1d6c46cda1fd
23:49:50 Searching Annoy index using 1 thread, search_k = 3000
23:49:50 Annoy recall = 100%
23:49:51 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
23:49:51 Initializing from normalized Laplacian + noise (using RSpectra)
23:49:51 Commencing optimization for 500 epochs, with 55346 positive edges
23:49:52 Optimization finished</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>(koala_umap_plot <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(koala_umap, <span class="at">reduction =</span> <span class="st">'umap'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="identify-cluster-biomarkers" class="level2">
<h2 class="anchored" data-anchor-id="identify-cluster-biomarkers">Identify cluster biomarkers</h2>
<section id="cluster-3" class="level3">
<h3 class="anchored" data-anchor-id="cluster-3">Cluster 3</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>koala_umap_cluster_4 <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(koala_umap, <span class="at">ident.1 =</span> <span class="dv">4</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(koala_umap_cluster_4, <span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               p_val avg_log2FC pct.1 pct.2    p_val_adj
mt-Co2  7.527182e-31  1.2245951 0.985 0.670 1.311687e-26
Malat1  1.723282e-29 -0.7632667 1.000 0.999 3.002992e-25
mt-Co1  5.977587e-27  1.1122765 0.971 0.790 1.041654e-22
Gm42418 3.151082e-26  0.5404391 1.000 1.000 5.491075e-22
Camk1d  3.573460e-26  0.8369697 0.993 0.950 6.227112e-22
Cmss1   5.424007e-26  0.8626727 0.993 0.932 9.451874e-22
Apoe    6.872890e-26  1.2087947 0.919 0.635 1.197670e-21
mt-Nd2  8.808615e-25  1.2449740 0.897 0.524 1.534989e-20
Stab2   1.176373e-24  0.8467994 0.875 0.327 2.049947e-20
mt-Atp6 1.815371e-24  1.1416188 0.949 0.667 3.163465e-20</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># find all markers distinguishing cluster 5 from clusters 0 through 3</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>koala_umap_cluster_4_diff <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(koala_umap, <span class="at">ident.1 =</span> <span class="dv">4</span>, <span class="at">ident.2 =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>)</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(koala_umap_cluster_4_diff, <span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               p_val avg_log2FC pct.1 pct.2    p_val_adj
mt-Co2  9.312486e-27  1.1041138 0.985 0.678 1.622794e-22
Malat1  2.704755e-26 -0.7462688 1.000 0.999 4.713307e-22
Gm42418 3.811073e-26  0.5477534 1.000 1.000 6.641176e-22
Camk1d  1.292259e-25  0.8488007 0.993 0.945 2.251891e-21
Apoe    9.025020e-25  1.1895949 0.919 0.622 1.572700e-20
Cmss1   2.670717e-24  0.8297866 0.993 0.925 4.653991e-20
mt-Co1  7.586908e-23  0.9905739 0.971 0.803 1.322095e-18
Lars2   1.597419e-22  0.8533974 1.000 0.832 2.783662e-18
mt-Atp6 2.901863e-20  0.9995626 0.949 0.680 5.056787e-16
mt-Nd2  3.537213e-20  1.0947916 0.897 0.557 6.163948e-16</code></pre>
</div>
</div>
</section>
<section id="all-marker-analysis" class="level3">
<h3 class="anchored" data-anchor-id="all-marker-analysis">All marker analysis</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># find markers for every cluster compared to all other cells, keep only the positive ones</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>koala_umap_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(koala_umap, <span class="at">only.pos =</span> <span class="cn">TRUE</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 0</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 5</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 6</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 7</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_10_per_cluster <span class="ot">&lt;-</span> koala_umap_markers <span class="sc">%&gt;%</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">order_by =</span> avg_log2FC)</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_10_per_cluster</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 90 × 7
# Groups:   cluster [9]
       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene    
       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;   
 1 6.94e- 80       3.90 0.488 0.066 1.21e- 75 0       Bmp6    
 2 2.11e- 46       3.86 0.281 0.03  3.68e- 42 0       Col13a1 
 3 8.78e- 66       3.67 0.406 0.052 1.53e- 61 0       Galnt18 
 4 1.10e-126       3.61 0.73  0.113 1.93e-122 0       Adam23  
 5 2.37e- 48       3.60 0.313 0.04  4.13e- 44 0       Dnase1l3
 6 5.07e- 46       3.56 0.294 0.035 8.83e- 42 0       Srgap1  
 7 6.73e- 42       3.52 0.283 0.039 1.17e- 37 0       Clec4g  
 8 6.99e-172       3.49 0.929 0.186 1.22e-167 0       Stab2   
 9 4.34e- 77       3.40 0.488 0.069 7.57e- 73 0       Rasgrp3 
10 4.35e- 46       3.34 0.322 0.048 7.59e- 42 0       Gm15675 
# ℹ 80 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_10_per_nest <span class="ot">&lt;-</span> koala_umap_markers_10_per_cluster <span class="sc">%&gt;%</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">nest</span>() <span class="sc">%&gt;%</span></span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">data =</span> <span class="fu">as.list</span>(<span class="fu">map</span>(data, <span class="sc">~</span>.x[[<span class="dv">1</span>]] )))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding missing grouping variables: `cluster`</code></pre>
</div>
</div>
</section>
<section id="set-up-function-for-all" class="level3">
<h3 class="anchored" data-anchor-id="set-up-function-for-all">Set up function for all</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>go_query_fun <span class="ot">&lt;-</span> <span class="cf">function</span>(gene_name_vector){</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>  go_result <span class="ot">&lt;-</span> <span class="fu">gost</span>(<span class="at">query =</span> gene_name_vector,</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">organism =</span> <span class="st">"mmusculus"</span>, <span class="at">ordered_query =</span> <span class="cn">FALSE</span>, </span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">multi_query =</span> <span class="cn">FALSE</span>, <span class="at">significant =</span> <span class="cn">TRUE</span>, <span class="at">exclude_iea =</span> <span class="cn">FALSE</span>, </span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">measure_underrepresentation =</span> <span class="cn">FALSE</span>, <span class="at">evcodes =</span> <span class="cn">FALSE</span>, </span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">user_threshold =</span> <span class="fl">0.05</span>, <span class="at">correction_method =</span> <span class="st">"fdr"</span>, </span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">domain_scope =</span> <span class="st">"annotated"</span>, <span class="at">custom_bg =</span> <span class="cn">NULL</span>, </span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">numeric_ns =</span> <span class="st">""</span>, <span class="at">sources =</span> <span class="st">"GO:MF"</span>, <span class="at">as_short_link =</span> <span class="cn">FALSE</span>)</span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(go_result)</span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-the-functon-in-df-env" class="level3">
<h3 class="anchored" data-anchor-id="run-the-functon-in-df-env">Run the functon in df env</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_10_per_nest_go_ano <span class="ot">&lt;-</span> koala_umap_markers_10_per_nest <span class="sc">%&gt;%</span> </span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">"go_table"</span> <span class="ot">=</span> <span class="fu">map</span>(data, go_query_fun))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>No results to show
Please make sure that the organism is correct or set significant = FALSE
No results to show
Please make sure that the organism is correct or set significant = FALSE</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_10_go_df <span class="ot">&lt;-</span> koala_umap_markers_10_per_nest_go_ano <span class="sc">%&gt;%</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(go_table) <span class="sc">%&gt;%</span></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">row_number</span>() <span class="sc">%%</span> <span class="dv">2</span> <span class="sc">!=</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(go_table) <span class="sc">%&gt;%</span></span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(term_name, <span class="at">.after =</span> cluster) <span class="sc">%&gt;%</span></span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(cluster, p_value) <span class="sc">%&gt;%</span></span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">!</span><span class="fu">c</span>(<span class="st">"data"</span>, <span class="st">"query"</span>, <span class="st">"significant"</span>))</span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(koala_umap_markers_10_go_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 13
# Groups:   cluster [1]
  cluster term_name     p_value term_size query_size intersection_size precision
  &lt;fct&gt;   &lt;chr&gt;           &lt;dbl&gt;     &lt;int&gt;      &lt;int&gt;             &lt;int&gt;     &lt;dbl&gt;
1 0       virus corece…  0.0317         1         10                 1       0.1
2 0       polysacchari…  0.0453        26         10                 1       0.1
3 0       protein-lipi…  0.0453        30         10                 1       0.1
4 0       lipoprotein …  0.0453        30         10                 1       0.1
5 0       BMP receptor…  0.0453        14         10                 1       0.1
6 0       transmembran…  0.0453        21         10                 1       0.1
# ℹ 6 more variables: recall &lt;dbl&gt;, term_id &lt;chr&gt;, source &lt;chr&gt;,
#   effective_domain_size &lt;int&gt;, source_order &lt;int&gt;, parents &lt;list&gt;</code></pre>
</div>
</div>
</section>
<section id="all-marker-analysis-1" class="level3">
<h3 class="anchored" data-anchor-id="all-marker-analysis-1">All marker analysis</h3>
<p>without requiring uniqueness</p>
<p>Let’s run the analysis without requiring Variable genes be uniquely upregulated or downregulated in each cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># find markers for every cluster, keep only the positive ones</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>koala_umap_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(koala_umap, <span class="at">only.pos =</span> <span class="cn">FALSE</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 0</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 5</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 6</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 7</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_10_per_cluster <span class="ot">&lt;-</span> koala_umap_markers <span class="sc">%&gt;%</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">order_by =</span> avg_log2FC)</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_10_per_cluster</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 90 × 7
# Groups:   cluster [9]
       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene    
       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;   
 1 6.94e- 80       3.90 0.488 0.066 1.21e- 75 0       Bmp6    
 2 2.11e- 46       3.86 0.281 0.03  3.68e- 42 0       Col13a1 
 3 8.78e- 66       3.67 0.406 0.052 1.53e- 61 0       Galnt18 
 4 1.10e-126       3.61 0.73  0.113 1.93e-122 0       Adam23  
 5 2.37e- 48       3.60 0.313 0.04  4.13e- 44 0       Dnase1l3
 6 5.07e- 46       3.56 0.294 0.035 8.83e- 42 0       Srgap1  
 7 6.73e- 42       3.52 0.283 0.039 1.17e- 37 0       Clec4g  
 8 6.99e-172       3.49 0.929 0.186 1.22e-167 0       Stab2   
 9 4.34e- 77       3.40 0.488 0.069 7.57e- 73 0       Rasgrp3 
10 4.35e- 46       3.34 0.322 0.048 7.59e- 42 0       Gm15675 
# ℹ 80 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_10_per_nest <span class="ot">&lt;-</span> koala_umap_markers_10_per_cluster <span class="sc">%&gt;%</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">nest</span>() <span class="sc">%&gt;%</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">data =</span> <span class="fu">as.list</span>(<span class="fu">map</span>(data, <span class="sc">~</span>.x[[<span class="dv">1</span>]] )))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding missing grouping variables: `cluster`</code></pre>
</div>
</div>
</section>
<section id="run-the-functon-in-df-env-1" class="level3">
<h3 class="anchored" data-anchor-id="run-the-functon-in-df-env-1">Run the functon in df env</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_10_per_nest_go_ano <span class="ot">&lt;-</span> koala_umap_markers_10_per_nest <span class="sc">%&gt;%</span> </span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">"go_table"</span> <span class="ot">=</span> <span class="fu">map</span>(data, go_query_fun))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>No results to show
Please make sure that the organism is correct or set significant = FALSE
No results to show
Please make sure that the organism is correct or set significant = FALSE</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>koala_umap_2way_markers_10_go_df <span class="ot">&lt;-</span> koala_umap_markers_10_per_nest_go_ano <span class="sc">%&gt;%</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(go_table) <span class="sc">%&gt;%</span></span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">row_number</span>() <span class="sc">%%</span> <span class="dv">2</span> <span class="sc">!=</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(go_table) <span class="sc">%&gt;%</span></span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(term_name, <span class="at">.after =</span> cluster) <span class="sc">%&gt;%</span></span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(cluster, <span class="fu">desc</span>(p_value))</span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(koala_umap_markers_10_go_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 13
# Groups:   cluster [1]
  cluster term_name     p_value term_size query_size intersection_size precision
  &lt;fct&gt;   &lt;chr&gt;           &lt;dbl&gt;     &lt;int&gt;      &lt;int&gt;             &lt;int&gt;     &lt;dbl&gt;
1 0       virus corece…  0.0317         1         10                 1       0.1
2 0       polysacchari…  0.0453        26         10                 1       0.1
3 0       protein-lipi…  0.0453        30         10                 1       0.1
4 0       lipoprotein …  0.0453        30         10                 1       0.1
5 0       BMP receptor…  0.0453        14         10                 1       0.1
6 0       transmembran…  0.0453        21         10                 1       0.1
# ℹ 6 more variables: recall &lt;dbl&gt;, term_id &lt;chr&gt;, source &lt;chr&gt;,
#   effective_domain_size &lt;int&gt;, source_order &lt;int&gt;, parents &lt;list&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(koala_umap_2way_markers_10_go_df, <span class="st">"koala_umap_2way_markers_10_go_bdf.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>koala_umap_cluster_7 <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(koala_umap, <span class="at">ident.1 =</span> <span class="dv">7</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>)</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(koala_umap_cluster_7, <span class="at">n =</span> <span class="dv">10</span>)[<span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>data frame with 0 columns and 10 rows</code></pre>
</div>
</div>
</section>
<section id="cluster-7" class="level3">
<h3 class="anchored" data-anchor-id="cluster-7">Cluster 7</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>koala_umap_cluster_unique_7 <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(koala_umap, <span class="at">ident.1 =</span> <span class="dv">7</span>, <span class="at">ident.2 =</span> (<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>)[<span class="sc">!</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>) <span class="sc">==</span> <span class="dv">7</span>], <span class="at">min.pct =</span> <span class="fl">0.5</span>)</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">as_tibble</span>(koala_umap_cluster_unique_7, <span class="at">rownames =</span> <span class="st">"gene"</span>), <span class="at">n =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 6
   gene        p_val avg_log2FC pct.1 pct.2 p_val_adj
   &lt;chr&gt;       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;
 1 Mecom    2.22e-75       7.23 0.5   0.011  3.86e-71
 2 Jam2     5.96e-62       4.89 0.587 0.029  1.04e-57
 3 Heg1     4.21e-42       4.27 0.652 0.071  7.33e-38
 4 Cyyr1    3.32e-41       4.26 0.674 0.08   5.78e-37
 5 Pecam1   1.37e-38       4.35 0.587 0.061  2.39e-34
 6 Ptprb    1.13e-35       2.97 0.891 0.181  1.96e-31
 7 Plcb4    5.14e-32       3.71 0.543 0.063  8.96e-28
 8 Meis2    9.37e-32       3.24 0.717 0.122  1.63e-27
 9 Cdh13    1.61e-30       4.63 0.5   0.058  2.81e-26
10 Prex2    7.70e-28       3.09 0.761 0.176  1.34e-23
11 Cemip2   2.88e-27       3.52 0.674 0.126  5.02e-23
12 Crim1    2.23e-26       3.40 0.652 0.129  3.88e-22
13 Adgrf5   3.18e-25       3.29 0.522 0.076  5.54e-21
14 Calcrl   9.21e-21       2.85 0.609 0.132  1.61e-16
15 Arhgap31 8.63e-20       3.32 0.5   0.092  1.50e-15
16 Il6st    3.91e-18       2.71 0.652 0.183  6.82e-14
17 Eng      6.85e-18       2.88 0.5   0.099  1.19e-13
18 Ptprm    1.08e-15       2.31 0.587 0.152  1.88e-11
19 Plekhg1  2.40e-15       2.48 0.5   0.113  4.18e-11
20 Arl15    6.65e-14       2.69 0.717 0.325  1.16e- 9</code></pre>
</div>
</div>
<p>## postsynaptic density protein 95 clustering 6 2 .00 &gt; 100 + 5.20E-06 8.18E-03 ## postsynaptic membrane organization 31 2 .01 &gt; 100 + 9.74E-05 3.00E-02 ## postsynaptic density organization 28 2 .01 &gt; 100 + 8.03E-05 2.81E-02 ## postsynaptic specialization organization 34 2 .02 &gt; 100 + 1.16E-04 3.45E-02 ## NMDA glutamate receptor clustering 6 2 .00 &gt; 100 + 5.20E-06 7.43E-03 ## neurotransmitter-gated ion channel clustering 15 2 .01 &gt; 100 + 2.52E-05 1.80E-02 ## positive regulation of synapse maturation 10 2 .00 &gt; 100 + 1.22E-05 1.38E-02 ## regulation of synapse maturation 23 2 .01 &gt; 100 + 5.54E-05 2.64E-02</p>
<p>Cluster 7 looks like it may be</p>
</section>
</section>
</section>
<section id="cell-type-assignments" class="level1">
<h1>Cell type assignments</h1>
<section id="using-garnett" class="level2">
<h2 class="anchored" data-anchor-id="using-garnett">Using garnett</h2>
<p><a href="https://%20cole-trapnell-lab.github.io/garnett">Garnett</a> is a semi-supervised method of cell-type assignment for single cell data that is compatible with Seurat objects. It consists of four components. First, Garnett defines a markup language for specifying cell types using the genes that they specifically express. The markup language is hierarchical in that a cell type can have subtypes (for example, CD4+ and CD8+ are subsets of T cells). Second, Garnett includes a parser that processes the markup file together with a single-cell dataset, identifying representative cells bearing markers that unambiguously identify them as one of the cell types defined in the file. Third, Garnett trains a classifier that recognizes additional cells as belonging to each cell type based on their similarity to representative cells, similar to an approach that our groups recently developed for annotating a single-cell mouse atlas of chromatin accessibility. Garnett does not require that cells be organized into clusters, but it can optionally extend classifications to additional cells using either its own internal clustering routines or those of other tools. Finally, Garnett provides a method for applying a classifier trained on one dataset to rapidly annotate additional datasets.</p>
<section id="classifier" class="level3">
<h3 class="anchored" data-anchor-id="classifier">Classifier</h3>
<p><a href="https://cole-trapnell-lab.github.io/garnett/classifiers/">Garnett’s directory of already-available classifiers</a> indicates that one is available for Mouse Brain and spinal cord tissues, which I hypothesize is the primary content of this sample I am analyzed based on the gene ontology terms displayed and remarked upon in the previous section. The data are from Zeisel’s and Linnarrson’s <em>et al.</em> <a href="https://www.sciencedirect.com/science/article/pii/S009286741830789X">“Molecular Architecture of the Mouse Nervous System”</a> and were trained and deposited by <a href="https://www.nature.com/articles/s41592-019-0535-3">Pliner &amp; Trapnell <em>et al.</em></a>.</p>
<!-- I will download the "mmbrain" RDS file and import it into R as a classifier—which doesn't require a import special function; all R object structed is saved when the generic `saveRDS()` export function is run. -->
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download.file("https://cole-trapnell-lab.github.io/garnett/classifiers/mmBrain_20191017.RDS",</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="co">#               "mmBrain_20191017.RDS")</span></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://cole-trapnell-lab.github.io/garnett/marker_files/mmBrain_markers.txt"</span>,</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>              <span class="st">"mmBrain_markers.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="prepare-scrna-seq-data-for-garnett" class="level3">
<h3 class="anchored" data-anchor-id="prepare-scrna-seq-data-for-garnett">Prepare scRNA-seq data for garnett</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Seurat to convert Seurat dimensionally reduced object to a CellDataSet object</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>koala_umap_cds <span class="ot">&lt;-</span> <span class="fu">as.CellDataSet</span>(koala_umap)</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a><span class="co"># generate size factors for normalization later</span></span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>koala_umap_cds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(koala_umap_cds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="prepare-trained-classified-from-heirarchical-marker-gene-file" class="level3">
<h3 class="anchored" data-anchor-id="prepare-trained-classified-from-heirarchical-marker-gene-file">Prepare trained classified from heirarchical marker gene file</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Mm.eg.db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: AnnotationDbi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:Matrix':

    expand, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:tidyr':

    expand</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:dplyr':

    first, rename</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'IRanges'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:sp':

    %over%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:purrr':

    reduce</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:dplyr':

    collapse, desc, slice</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'AnnotationDbi'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    select</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>marker_file_path <span class="ot">&lt;-</span> <span class="st">"mmBrain_markers.txt"</span></span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a>marker_check <span class="ot">&lt;-</span> <span class="fu">check_markers</span>(koala_umap_cds, marker_file_path,</span>
<span id="cb123-4"><a href="#cb123-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">db =</span> org.Mm.eg.db,</span>
<span id="cb123-5"><a href="#cb123-5" aria-hidden="true" tabindex="-1"></a>                              <span class="at">cds_gene_id_type =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb123-6"><a href="#cb123-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">marker_file_gene_id_type =</span> <span class="st">"SYMBOL"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>There are 144 cell type definitions</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>marker_cell_types_analysis <span class="ot">&lt;-</span> marker_check <span class="sc">%&gt;%</span></span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">"n_type"</span> <span class="ot">=</span> <span class="fu">n</span>(), <span class="at">.by =</span> <span class="fu">c</span>(<span class="st">"cell_type"</span>, <span class="st">"parent"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">"n_cds"</span> <span class="ot">=</span> <span class="fu">n</span>(), <span class="at">.by =</span> <span class="fu">c</span>(<span class="st">"cell_type"</span>, <span class="st">"parent"</span>, <span class="st">"in_cds"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(n_type)) <span class="sc">%&gt;%</span></span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">c</span>(<span class="st">"cell_type"</span>, <span class="st">"parent"</span>, <span class="st">"in_cds"</span>, <span class="st">"n_type"</span>, <span class="st">"n_cds"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb125-6"><a href="#cb125-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>()</span>
<span id="cb125-7"><a href="#cb125-7" aria-hidden="true" tabindex="-1"></a>marker_cell_types_analysis</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                                             cell_type
1                                                              Neurons
2                                                              Neurons
3                    Cholinergic monoaminergic and peptidergic neurons
4                    Cholinergic monoaminergic and peptidergic neurons
5                                                             CNS glia
6                                                             CNS glia
7                                        Di- and mesencephalon neurons
8                                        Di- and mesencephalon neurons
9                                           Telencephalon interneurons
10                                          Telencephalon interneurons
11                               Cholinergic and monoaminergic neurons
12                               Cholinergic and monoaminergic neurons
13                                    Telencephalon projecting neurons
14                                    Telencephalon projecting neurons
15                                                Astroependymal cells
16                                                Astroependymal cells
17                                          Peripheral sensory neurons
18                                          Peripheral sensory neurons
19                               Telencephalon inhibitory interneurons
20                               Telencephalon inhibitory interneurons
21                            Di- and mesencephalon excitatory neurons
22                            Di- and mesencephalon excitatory neurons
23                                              Neural crest-like glia
24                                              Neural crest-like glia
25                                                 Peptidergic neurons
26                                                 Peptidergic neurons
27                                                    Oligodendrocytes
28                                                    Oligodendrocytes
29                                                      Vascular cells
30                                                      Vascular cells
31                         Telencephalon projecting excitatory neurons
32                         Telencephalon projecting excitatory neurons
33                            Di- and mesencephalon inhibitory neurons
34                            Di- and mesencephalon inhibitory neurons
35                                                          Astrocytes
36                                                          Astrocytes
37                          Peripheral sensory non-peptidergic neurons
38                          Peripheral sensory non-peptidergic neurons
39                                                     Enteric neurons
40                                                     Enteric neurons
41                                  Excitatory neurons cerebral cortex
42                                  Excitatory neurons cerebral cortex
43                                                 Spinal cord neurons
44                                                 Spinal cord neurons
45                         Telencephalon projecting inhibitory neurons
46                         Telencephalon projecting inhibitory neurons
47                                                        Enteric glia
48                                                        Enteric glia
49                                                     Immature neural
50                                                     Immature neural
51                              Peripheral sensory peptidergic neurons
52                              Peripheral sensory peptidergic neurons
53                                        Olfactory inhibitory neurons
54                                        Olfactory inhibitory neurons
55                                      Inhibitory neurons midbrain di
56                                      Inhibitory neurons midbrain di
57                                                Serotonergic neurons
58                                                Serotonergic neurons
59                                                        Immune cells
60                                                        Immune cells
61                                                 Sympathetic neurons
62                                                 Sympathetic neurons
63                                                   Hindbrain neurons
64                                                   Hindbrain neurons
65                                         Cholinergic enteric neurons
66                                         Cholinergic enteric neurons
67                                                  Cerebellum neurons
68                                                  Cerebellum neurons
69                                                           Microglia
70                                                           Microglia
71                                         Excitatory neurons midbrain
72                                         Excitatory neurons midbrain
73                                             Mature oligodendrocytes
74                                             Mature oligodendrocytes
75                                      Spinal cord excitatory neurons
76                                      Spinal cord excitatory neurons
77                                                     Ependymal cells
78                                                     Ependymal cells
79                                           Glutamatergic neuroblasts
80                                           Glutamatergic neuroblasts
81                                   Sympathetic noradrenergic neurons
82                                   Sympathetic noradrenergic neurons
83                                                           Pericytes
84                                                           Pericytes
85                            Peripheral sensory neurofilament neurons
86                            Peripheral sensory neurofilament neurons
87                                       Vascular leptomeningeal cells
88                                       Vascular leptomeningeal cells
89                                    Peptidergic neurons hypothalamus
90                                    Peptidergic neurons hypothalamus
91                                       Non-glutamatergic neuroblasts
92                                       Non-glutamatergic neuroblasts
93                                      Spinal cord inhibitory neurons
94                                      Spinal cord inhibitory neurons
95                     R-LM border Cck interneurons cortex/hippocampus
96                                   Inhibitory neurons olfactory bulb
97                                   Inhibitory neurons olfactory bulb
98                                      Excitatory neurons spinal cord
99                                    D1 medium spiny neurons striatum
100                                   D1 medium spiny neurons striatum
101                                    Myelin forming oligodendrocytes
102                                    Myelin forming oligodendrocytes
103                               Noradrenergic neurons of the medulla
104                               Noradrenergic neurons of the medulla
105                                          Nitrergic enteric neurons
106                                          Nitrergic enteric neurons
107                                  Excitatory neurons hindbrain hind
108                                  Excitatory neurons hindbrain hind
109                                              Neuroblasts olfactory
110                                       Non-telencephalon astrocytes
111                                       Non-telencephalon astrocytes
112                                Inhibitory neurons spinal cord spin
113                                                     Satellite glia
114                                                     Satellite glia
115                                      Newly formed oligodendrocytes
116                                      Newly formed oligodendrocytes
117                                      Dentate gyrus granule neurons
118                                      Dentate gyrus granule neurons
119 Sleep-active long-range projection interneurons cortex/hippocampus
120 Sleep-active long-range projection interneurons cortex/hippocampus
121                                    Inhibitory neurons hindbrain di
122                                    Inhibitory neurons hindbrain di
123                                       Vascular smooth muscle cells
124                                       Vascular smooth muscle cells
125                                    Excitatory neurons hypothalamus
126                                                     Purkinje cells
127                                                     Purkinje cells
128                           Afferent nuclei of cranial nerves VI-XII
129                           Afferent nuclei of cranial nerves VI-XII
130                                           Perivascular macrophages
131                                           Perivascular macrophages
132                           Cholinergic neurons midbrain red nucleus
133                                   Inhibitory neurons telencephalon
134                     Non-border Cck interneurons cortex/hippocampus
135                                  Central canal neurons spinal cord
136                                           Choroid epithelial cells
137                                    Enteric mesothelial fibroblasts
138                           Dopaminergic neurons periaqueductal grey
139                External plexiform layer interneuron olfactory bulb
140                              Adrenergic cell groups of the medulla
141                              Adrenergic cell groups of the medulla
142                                        Excitatory neurons thalamus
143                                        Excitatory neurons thalamus
144              Interneuron-selective interneurons cortex/hippocampus
145              Interneuron-selective interneurons cortex/hippocampus
146                              Subcommissural organ hypendymal cells
147                                           Telencephalon astrocytes
148                                   D2 medium spiny neurons striatum
149                                   D2 medium spiny neurons striatum
150                                  Excitatory neurons hindbrain spin
151                                      Granule neurons dentate gyrus
152                                      Granule neurons dentate gyrus
153                                    Sympathetic cholinergic neurons
154                                    Sympathetic cholinergic neurons
155                                   Committed oligodendrocytes cells
156                                   Committed oligodendrocytes cells
157                                        Olfactory ensheathing cells
158                                        Olfactory ensheathing cells
159                                       Excitatory neurons hindbrain
160                                         Vascular endothelial cells
161                               Inner horizontal cell olfactory bulb
162                               Inner horizontal cell olfactory bulb
163                         Subventricular zone radial glia-like cells
164                         Subventricular zone radial glia-like cells
165                                  Inhibitory neurons hindbrain hind
166                                  Inhibitory neurons hindbrain hind
167                                         Matrix D1 neurons striatum
168                                                      Bergmann Glia
169                 CGE-derived neurogliaform cells cortex/hippocampus
170                                        Neuroblasts cerebellum glut
171                                    Cajal-Retzius cells hippocampus
172                               Dentate gyrus radial glia-like cells
173              Glutamatergic projection neurons of the raphe nucleus
174                                   Inhibitory neurons midbrain pept
175                                  Granule neuroblasts dentate gyrus
176                                      Cholinergic neurons hindbrain
177                                       Patch D1/D2 neurons striatum
178     Cholinergic neurons septal nucleus Meissnert and diagonal band
179                                 Inhibitory neurons hypothalamus di
180                                         Granule neurons cerebellum
181                              Neuronal intermidate progenitor cells
182                                               Olfactory astrocytes
183                                         Neuroblasts olfactory bulb
184                                   Oligodendrocytes precursor cells
185                                        Inhibitory neurons thalamus
186                                Inhibitory interneurons hippocampus
187                             Granule neuroblasts dentate gyrus dent
188                                        Inhibitory neurons midbrain
189                                                      Schwann cells
190                                             Neuroblasts cerebellum
191                            Molecular layer interneurons cerebellum
192                                           Neuroblast-like habenula
193                              Oxytocin-producing cells hypothalamus
194                                          Pmch neurons hypothalamus
195                           Vasopressin-producing cells hypothalamus
196         Ivy and MGE-derived neurogliaform cells cortex/hippocampus
197                                                 Neuroblasts septum
198                                     Inhibitory neurons spinal cord
199                                 Excitatory neurons hippocampus CA1
200                   Basket and bistratified cells cortex/hippocampus
201                                 Excitatory neurons hippocampus CA3
202                             Cholinergic interneurons telencephalon
203                            Afferent nuclei of cranial nerves III-V
204                                        Inhibitory neurons pallidum
205                                   Inhibitory neurons thalamus tele
206                                    Inhibitory neurons hypothalamus
207             Dopaminergic periglomerular interneuron olfactory bulb
208                     Hippocamposeptal projection cortex/hippocampus
209                                       Trilaminar cells hippocampus
210                              Dopaminergic neurons ventral midbrain
211                             Granular layer interneurons cerebellum
212                                      Axo-axonic cortex/hippocampus
213                     Dorsal midbrain Myoc-expressing astrocyte-like
214                                        Excitatory neurons amygdala
215                              Orexin-producing neurons hypothalamus
216                                  Inhibitory neurons septal nucleus
217                                       Inhibitory neurons hindbrain
218                                       Cholinergic neurons habenula
                                               parent in_cds n_type n_cds
1                                                root  FALSE    197   119
2                                                root   TRUE    197    78
3                                             Neurons  FALSE     42    26
4                                             Neurons   TRUE     42    16
5                                                root   TRUE     38    17
6                                                root  FALSE     38    21
7                                             Neurons  FALSE     31    19
8                                             Neurons   TRUE     31    12
9                                             Neurons  FALSE     28    18
10                                            Neurons   TRUE     28    10
11  Cholinergic monoaminergic and peptidergic neurons   TRUE     25    10
12  Cholinergic monoaminergic and peptidergic neurons  FALSE     25    15
13                                            Neurons  FALSE     25    14
14                                            Neurons   TRUE     25    11
15                                           CNS glia   TRUE     24    11
16                                           CNS glia  FALSE     24    13
17                                            Neurons  FALSE     23    14
18                                            Neurons   TRUE     23     9
19                         Telencephalon interneurons  FALSE     20    12
20                         Telencephalon interneurons   TRUE     20     8
21                      Di- and mesencephalon neurons  FALSE     18    14
22                      Di- and mesencephalon neurons   TRUE     18     4
23                                               root   TRUE     18    12
24                                               root  FALSE     18     6
25  Cholinergic monoaminergic and peptidergic neurons  FALSE     16    10
26  Cholinergic monoaminergic and peptidergic neurons   TRUE     16     6
27                                           CNS glia  FALSE     14     8
28                                           CNS glia   TRUE     14     6
29                                               root  FALSE     13     5
30                                               root   TRUE     13     8
31                   Telencephalon projecting neurons   TRUE     13     5
32                   Telencephalon projecting neurons  FALSE     13     8
33                      Di- and mesencephalon neurons   TRUE     13     8
34                      Di- and mesencephalon neurons  FALSE     13     5
35                               Astroependymal cells  FALSE     12     5
36                               Astroependymal cells   TRUE     12     7
37                         Peripheral sensory neurons  FALSE     11     9
38                         Peripheral sensory neurons   TRUE     11     2
39                                            Neurons  FALSE     10     7
40                                            Neurons   TRUE     10     3
41        Telencephalon projecting excitatory neurons   TRUE     10     4
42        Telencephalon projecting excitatory neurons  FALSE     10     6
43                                            Neurons  FALSE     10     5
44                                            Neurons   TRUE     10     5
45                   Telencephalon projecting neurons  FALSE      9     4
46                   Telencephalon projecting neurons   TRUE      9     5
47                             Neural crest-like glia  FALSE      9     3
48                             Neural crest-like glia   TRUE      9     6
49                                            Neurons  FALSE      9     5
50                                            Neurons   TRUE      9     4
51                         Peripheral sensory neurons  FALSE      8     3
52                         Peripheral sensory neurons   TRUE      8     5
53                         Telencephalon interneurons  FALSE      8     6
54                         Telencephalon interneurons   TRUE      8     2
55           Di- and mesencephalon inhibitory neurons  FALSE      8     3
56           Di- and mesencephalon inhibitory neurons   TRUE      8     5
57              Cholinergic and monoaminergic neurons   TRUE      8     5
58              Cholinergic and monoaminergic neurons  FALSE      8     3
59                                               root   TRUE      8     4
60                                               root  FALSE      8     4
61                                            Neurons  FALSE      7     3
62                                            Neurons   TRUE      7     4
63                                            Neurons  FALSE      7     5
64                                            Neurons   TRUE      7     2
65                                    Enteric neurons   TRUE      7     2
66                                    Enteric neurons  FALSE      7     5
67                                            Neurons  FALSE      7     5
68                                            Neurons   TRUE      7     2
69                                       Immune cells   TRUE      6     3
70                                       Immune cells  FALSE      6     3
71           Di- and mesencephalon excitatory neurons  FALSE      6     4
72           Di- and mesencephalon excitatory neurons   TRUE      6     2
73                                   Oligodendrocytes  FALSE      6     4
74                                   Oligodendrocytes   TRUE      6     2
75                                Spinal cord neurons  FALSE      5     3
76                                Spinal cord neurons   TRUE      5     2
77                               Astroependymal cells  FALSE      5     4
78                               Astroependymal cells   TRUE      5     1
79                                    Immature neural  FALSE      5     4
80                                    Immature neural   TRUE      5     1
81                                Sympathetic neurons   TRUE      5     3
82                                Sympathetic neurons  FALSE      5     2
83                                     Vascular cells  FALSE      4     1
84                                     Vascular cells   TRUE      4     3
85                         Peripheral sensory neurons  FALSE      4     2
86                         Peripheral sensory neurons   TRUE      4     2
87                                     Vascular cells   TRUE      4     3
88                                     Vascular cells  FALSE      4     1
89                                Peptidergic neurons  FALSE      4     3
90                                Peptidergic neurons   TRUE      4     1
91                                    Immature neural   TRUE      4     3
92                                    Immature neural  FALSE      4     1
93                                Spinal cord neurons   TRUE      4     1
94                                Spinal cord neurons  FALSE      4     3
95              Telencephalon inhibitory interneurons   TRUE      3     3
96                       Olfactory inhibitory neurons   TRUE      3     1
97                       Olfactory inhibitory neurons  FALSE      3     2
98                     Spinal cord excitatory neurons  FALSE      3     3
99        Telencephalon projecting inhibitory neurons   TRUE      3     2
100       Telencephalon projecting inhibitory neurons  FALSE      3     1
101                                  Oligodendrocytes   TRUE      3     1
102                                  Oligodendrocytes  FALSE      3     2
103             Cholinergic and monoaminergic neurons  FALSE      3     1
104             Cholinergic and monoaminergic neurons   TRUE      3     2
105                                   Enteric neurons  FALSE      3     2
106                                   Enteric neurons   TRUE      3     1
107                                 Hindbrain neurons  FALSE      3     2
108                                 Hindbrain neurons   TRUE      3     1
109                         Glutamatergic neuroblasts  FALSE      3     3
110                                        Astrocytes  FALSE      3     1
111                                        Astrocytes   TRUE      3     2
112                    Spinal cord inhibitory neurons  FALSE      3     3
113                            Neural crest-like glia  FALSE      3     2
114                            Neural crest-like glia   TRUE      3     1
115                                  Oligodendrocytes   TRUE      3     2
116                                  Oligodendrocytes  FALSE      3     1
117                  Telencephalon projecting neurons  FALSE      3     2
118                  Telencephalon projecting neurons   TRUE      3     1
119             Telencephalon inhibitory interneurons  FALSE      2     1
120             Telencephalon inhibitory interneurons   TRUE      2     1
121          Di- and mesencephalon inhibitory neurons   TRUE      2     1
122          Di- and mesencephalon inhibitory neurons  FALSE      2     1
123                                    Vascular cells  FALSE      2     1
124                                    Vascular cells   TRUE      2     1
125          Di- and mesencephalon excitatory neurons  FALSE      2     2
126                                Cerebellum neurons  FALSE      2     1
127                                Cerebellum neurons   TRUE      2     1
128             Cholinergic and monoaminergic neurons  FALSE      2     1
129             Cholinergic and monoaminergic neurons   TRUE      2     1
130                                      Immune cells  FALSE      2     1
131                                      Immune cells   TRUE      2     1
132          Di- and mesencephalon excitatory neurons  FALSE      2     2
133                               Peptidergic neurons  FALSE      2     2
134             Telencephalon inhibitory interneurons  FALSE      2     2
135                               Peptidergic neurons   TRUE      2     2
136                              Astroependymal cells  FALSE      2     2
137                            Neural crest-like glia   TRUE      2     2
138             Cholinergic and monoaminergic neurons  FALSE      2     2
139                      Olfactory inhibitory neurons  FALSE      2     2
140             Cholinergic and monoaminergic neurons   TRUE      2     1
141             Cholinergic and monoaminergic neurons  FALSE      2     1
142          Di- and mesencephalon excitatory neurons   TRUE      2     1
143          Di- and mesencephalon excitatory neurons  FALSE      2     1
144             Telencephalon inhibitory interneurons   TRUE      2     1
145             Telencephalon inhibitory interneurons  FALSE      2     1
146                              Astroependymal cells   TRUE      2     2
147                                        Astrocytes   TRUE      2     2
148       Telencephalon projecting inhibitory neurons   TRUE      2     1
149       Telencephalon projecting inhibitory neurons  FALSE      2     1
150                    Spinal cord excitatory neurons   TRUE      2     2
151                     Dentate gyrus granule neurons  FALSE      2     1
152                     Dentate gyrus granule neurons   TRUE      2     1
153                               Sympathetic neurons   TRUE      2     1
154                               Sympathetic neurons  FALSE      2     1
155                                  Oligodendrocytes  FALSE      2     1
156                                  Oligodendrocytes   TRUE      2     1
157                            Neural crest-like glia  FALSE      2     1
158                            Neural crest-like glia   TRUE      2     1
159          Di- and mesencephalon excitatory neurons  FALSE      2     2
160                                    Vascular cells  FALSE      2     2
161                      Olfactory inhibitory neurons  FALSE      2     1
162                      Olfactory inhibitory neurons   TRUE      2     1
163                              Astroependymal cells   TRUE      2     1
164                              Astroependymal cells  FALSE      2     1
165                                 Hindbrain neurons  FALSE      2     1
166                                 Hindbrain neurons   TRUE      2     1
167       Telencephalon projecting inhibitory neurons   TRUE      1     1
168                                        Astrocytes   TRUE      1     1
169             Telencephalon inhibitory interneurons  FALSE      1     1
170                         Glutamatergic neuroblasts   TRUE      1     1
171          Di- and mesencephalon excitatory neurons  FALSE      1     1
172                              Astroependymal cells  FALSE      1     1
173             Cholinergic and monoaminergic neurons  FALSE      1     1
174                               Peptidergic neurons  FALSE      1     1
175                     Non-glutamatergic neuroblasts  FALSE      1     1
176                                 Hindbrain neurons  FALSE      1     1
177       Telencephalon projecting inhibitory neurons   TRUE      1     1
178             Cholinergic and monoaminergic neurons  FALSE      1     1
179          Di- and mesencephalon inhibitory neurons   TRUE      1     1
180                                Cerebellum neurons  FALSE      1     1
181                     Non-glutamatergic neuroblasts   TRUE      1     1
182                                        Astrocytes   TRUE      1     1
183                     Non-glutamatergic neuroblasts   TRUE      1     1
184                            Neural crest-like glia   TRUE      1     1
185                               Peptidergic neurons   TRUE      1     1
186             Telencephalon inhibitory interneurons   TRUE      1     1
187                     Dentate gyrus granule neurons  FALSE      1     1
188                                Cerebellum neurons  FALSE      1     1
189                            Neural crest-like glia   TRUE      1     1
190                                Cerebellum neurons  FALSE      1     1
191                                Cerebellum neurons  FALSE      1     1
192                     Non-glutamatergic neuroblasts   TRUE      1     1
193                               Peptidergic neurons  FALSE      1     1
194             Cholinergic and monoaminergic neurons  FALSE      1     1
195                               Peptidergic neurons   TRUE      1     1
196             Telencephalon inhibitory interneurons   TRUE      1     1
197                         Glutamatergic neuroblasts  FALSE      1     1
198                                 Hindbrain neurons  FALSE      1     1
199       Telencephalon projecting excitatory neurons  FALSE      1     1
200             Telencephalon inhibitory interneurons  FALSE      1     1
201       Telencephalon projecting excitatory neurons  FALSE      1     1
202             Cholinergic and monoaminergic neurons  FALSE      1     1
203             Cholinergic and monoaminergic neurons  FALSE      1     1
204          Di- and mesencephalon inhibitory neurons  FALSE      1     1
205             Telencephalon inhibitory interneurons  FALSE      1     1
206                               Peptidergic neurons  FALSE      1     1
207                      Olfactory inhibitory neurons  FALSE      1     1
208             Telencephalon inhibitory interneurons  FALSE      1     1
209             Telencephalon inhibitory interneurons  FALSE      1     1
210             Cholinergic and monoaminergic neurons  FALSE      1     1
211                                Cerebellum neurons   TRUE      1     1
212             Telencephalon inhibitory interneurons  FALSE      1     1
213                                        Astrocytes  FALSE      1     1
214       Telencephalon projecting excitatory neurons   TRUE      1     1
215             Cholinergic and monoaminergic neurons  FALSE      1     1
216                               Peptidergic neurons  FALSE      1     1
217                    Spinal cord inhibitory neurons   TRUE      1     1
218          Di- and mesencephalon excitatory neurons  FALSE      1     1</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(marker_check_analysis, <span class="st">"mmBrain_markers_marker_check_analysis.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Make a table showing percent of genes for each cell type sub-classification that are present so I can filter and subset the marker genes to decrease the complexity of the model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>marker_cell_types_prefilter <span class="ot">&lt;-</span> marker_cell_types_analysis <span class="sc">%&gt;%</span></span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span>n_type) <span class="sc">%&gt;%</span></span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">id_cols =</span> <span class="fu">c</span>(<span class="st">"cell_type"</span>, <span class="st">"parent"</span>), <span class="at">values_from =</span> <span class="st">"n_cds"</span>,</span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">names_from =</span> <span class="st">"in_cds"</span>, <span class="at">names_glue =</span> <span class="st">"gene_present_{in_cds}"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">"percent_genes_present"</span> <span class="ot">=</span> gene_present_TRUE <span class="sc">/</span> (gene_present_TRUE <span class="sc">+</span> gene_present_FALSE)) <span class="sc">%&gt;%</span></span>
<span id="cb128-6"><a href="#cb128-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(gene_present_TRUE, percent_genes_present, <span class="at">.after =</span> parent) <span class="sc">%&gt;%</span></span>
<span id="cb128-7"><a href="#cb128-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(gene_present_TRUE), <span class="fu">desc</span>(percent_genes_present))</span>
<span id="cb128-8"><a href="#cb128-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(marker_cell_types_prefilter)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  cell_type    parent gene_present_TRUE percent_genes_present gene_present_FALSE
  &lt;chr&gt;        &lt;chr&gt;              &lt;int&gt;                 &lt;dbl&gt;              &lt;int&gt;
1 Neurons      root                  78                 0.396                119
2 CNS glia     root                  17                 0.447                 21
3 Cholinergic… Neuro…                16                 0.381                 26
4 Neural cres… root                  12                 0.667                  6
5 Di- and mes… Neuro…                12                 0.387                 19
6 Astroependy… CNS g…                11                 0.458                 13</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>marker_cell_types_postfilter <span class="ot">&lt;-</span> marker_cell_types_prefilter <span class="sc">%&gt;%</span></span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(parent <span class="sc">==</span> <span class="st">"root"</span> <span class="sc">|</span> gene_present_TRUE <span class="sc">&gt;</span> <span class="dv">8</span> <span class="sc">|</span> percent_genes_present <span class="sc">&gt;</span> <span class="fl">0.49</span>)</span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(marker_cell_types_postfilter)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  cell_type    parent gene_present_TRUE percent_genes_present gene_present_FALSE
  &lt;chr&gt;        &lt;chr&gt;              &lt;int&gt;                 &lt;dbl&gt;              &lt;int&gt;
1 Neurons      root                  78                 0.396                119
2 CNS glia     root                  17                 0.447                 21
3 Cholinergic… Neuro…                16                 0.381                 26
4 Neural cres… root                  12                 0.667                  6
5 Di- and mes… Neuro…                12                 0.387                 19
6 Astroependy… CNS g…                11                 0.458                 13</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>marker_df_filtered <span class="ot">&lt;-</span> marker_check <span class="sc">%&gt;%</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># filter for cell_types retained by filter line in "postfilter"</span></span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">semi_join</span>(marker_cell_types_postfilter, <span class="at">by =</span> <span class="st">"cell_type"</span>)</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(marker_df_filtered)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  marker_gene            gene_id                                parent
1       Crhbp ENSMUSG00000021680 Telencephalon inhibitory interneurons
2       Tacr1 ENSMUSG00000030043 Telencephalon inhibitory interneurons
3         Trh ENSMUSG00000005892                               Neurons
4        Trdn ENSMUSG00000019787                               Neurons
5      Nccrp1 ENSMUSG00000047586                               Neurons
6        Dlx5 ENSMUSG00000029755                               Neurons
                                                           cell_type in_cds
1 Sleep-active long-range projection interneurons cortex/hippocampus  FALSE
2 Sleep-active long-range projection interneurons cortex/hippocampus   TRUE
3                                         Telencephalon interneurons  FALSE
4                                         Telencephalon interneurons   TRUE
5                                         Telencephalon interneurons  FALSE
6                                         Telencephalon interneurons  FALSE
  nominates total_nominated exclusion_dismisses inclusion_ambiguates
1        NA              NA                  NA                   NA
2         1               1                   1                    0
3        NA              NA                  NA                   NA
4         1             158                   1                    1
5        NA              NA                  NA                   NA
6        NA              NA                  NA                   NA
        most_overlap ambiguity marker_score         summary
1               &lt;NA&gt;        NA           NA      Not in CDS
2               &lt;NA&gt;         0 1.000000e+02              Ok
3               &lt;NA&gt;        NA           NA      Not in CDS
4 Cerebellum neurons         1 6.266449e-03 High ambiguity?
5               &lt;NA&gt;        NA           NA      Not in CDS
6               &lt;NA&gt;        NA           NA      Not in CDS</code></pre>
</div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>marker_df_filtered_orphans <span class="ot">&lt;-</span> marker_df_filtered <span class="sc">%&gt;%</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">anti_join</span>(., ., <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"cell_type"</span> <span class="ot">=</span> <span class="st">"parent"</span>))</span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(marker_df_filtered_orphans)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  marker_gene            gene_id                                parent
1       Crhbp ENSMUSG00000021680 Telencephalon inhibitory interneurons
2       Tacr1 ENSMUSG00000030043 Telencephalon inhibitory interneurons
3         Trh ENSMUSG00000005892                               Neurons
4        Trdn ENSMUSG00000019787                               Neurons
5      Nccrp1 ENSMUSG00000047586                               Neurons
6        Dlx5 ENSMUSG00000029755                               Neurons
                                                           cell_type in_cds
1 Sleep-active long-range projection interneurons cortex/hippocampus  FALSE
2 Sleep-active long-range projection interneurons cortex/hippocampus   TRUE
3                                         Telencephalon interneurons  FALSE
4                                         Telencephalon interneurons   TRUE
5                                         Telencephalon interneurons  FALSE
6                                         Telencephalon interneurons  FALSE
  nominates total_nominated exclusion_dismisses inclusion_ambiguates
1        NA              NA                  NA                   NA
2         1               1                   1                    0
3        NA              NA                  NA                   NA
4         1             158                   1                    1
5        NA              NA                  NA                   NA
6        NA              NA                  NA                   NA
        most_overlap ambiguity marker_score         summary
1               &lt;NA&gt;        NA           NA      Not in CDS
2               &lt;NA&gt;         0 1.000000e+02              Ok
3               &lt;NA&gt;        NA           NA      Not in CDS
4 Cerebellum neurons         1 6.266449e-03 High ambiguity?
5               &lt;NA&gt;        NA           NA      Not in CDS
6               &lt;NA&gt;        NA           NA      Not in CDS</code></pre>
</div>
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a>marker_df_filtered_orph_rem <span class="ot">&lt;-</span>  <span class="fu">anti_join</span>(marker_df_filtered, marker_df_filtered_orphans, <span class="at">by =</span> <span class="st">"marker_gene"</span>)</span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(marker_df_filtered_orph_rem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  marker_gene            gene_id  parent                  cell_type in_cds
1      Gm7271               &lt;NA&gt; Neurons Peripheral sensory neurons  FALSE
2     Fam19a4               &lt;NA&gt; Neurons Peripheral sensory neurons  FALSE
3        Tlx3 ENSMUSG00000040610 Neurons Peripheral sensory neurons  FALSE
4      Mrgprd ENSMUSG00000051207 Neurons Peripheral sensory neurons  FALSE
5        Otoa ENSMUSG00000034990 Neurons Peripheral sensory neurons   TRUE
6     Mrgpra3 ENSMUSG00000078698 Neurons Peripheral sensory neurons  FALSE
  nominates total_nominated exclusion_dismisses inclusion_ambiguates
1        NA              NA                  NA                   NA
2        NA              NA                  NA                   NA
3        NA              NA                  NA                   NA
4        NA              NA                  NA                   NA
5         1             275                   0                    0
6        NA              NA                  NA                   NA
  most_overlap ambiguity marker_score    summary
1         &lt;NA&gt;        NA           NA  Not in db
2         &lt;NA&gt;        NA           NA  Not in db
3         &lt;NA&gt;        NA           NA Not in CDS
4         &lt;NA&gt;        NA           NA Not in CDS
5         &lt;NA&gt;         0    0.3636364         Ok
6         &lt;NA&gt;        NA           NA Not in CDS</code></pre>
</div>
</div>
<p>I also have to head off an error about parent cell types referred to in the parent field of other entries not being present by doing a recursive semi_join. See the error copied below: <code>Error in check_markers(koala_umap_cds, marker_file_path, db = org.Mm.eg.db, : Subtype Telencephalon inhibitory interneurons is not defined in marker file.Subtype Cerebellum neurons is not defined in marker file.Subtype root is not defined in marker file.Subtype Di- and mesencephalon excitatory neurons is not defined in marker file.Subtype Dentate gyrus granule neurons is not defined in marker file.Subtype Oligodendrocytes is not defined in marker file.Subtype Immature neural is not defined in marker file.Subtype Olfactory inhibitory neurons is not defined in marker file.Subtype Hindbrain neurons is not defined in marker file.</code></p>
<p><strong>2023-05-29 14:36 PM</strong> <strong>Where I left off: Marker checks are writing to disk and are parsing. But missing parents (orphaned cell types) are still in the marker .txt specs, and I have misplaced the dataframes that actually contain the gene info.</strong></p>
<section id="write-new-marker-file-to-disk" class="level4">
<h4 class="anchored" data-anchor-id="write-new-marker-file-to-disk">write new marker file to disk</h4>
<p>The basic structure of the Garnett marker file is a series of entries, each describing elements of a cell type. After the cell name, each additional line will be a descriptor, which begins with a keyword, followed by a colon (‘:’). After the colon, a series of specifications can be added, separated by commas (‘,’). Descriptors may spill onto following lines so long as you do not split a specification across multiple lines (i.e.&nbsp;if breaking up a long descriptor across multiple lines, all but the last line should end with a comma). Each new descriptor should begin on a new line. A generic cell type entry looks like this:</p>
<p>“’ &gt; cell type name descriptor: spec1, spec2, spec3, spec4 descriptor2: spec1 “’</p>
</section>
<section id="define-new-writer-function" class="level4">
<h4 class="anchored" data-anchor-id="define-new-writer-function">define new writer function</h4>
<p>The joy of for loops</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>write_garnett_marker_file <span class="ot">&lt;-</span> <span class="cf">function</span>(marker_df, filename){</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a>  marker_df <span class="ot">&lt;-</span> marker_df <span class="sc">%&gt;%</span></span>
<span id="cb138-4"><a href="#cb138-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(cell_type, parent, marker_gene) <span class="sc">%&gt;%</span></span>
<span id="cb138-5"><a href="#cb138-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="st">"marker_gene"</span> <span class="ot">=</span> <span class="fu">paste</span>(marker_gene, <span class="at">collapse =</span> <span class="st">", "</span>),</span>
<span id="cb138-6"><a href="#cb138-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">.by =</span> <span class="fu">c</span>(cell_type, parent))</span>
<span id="cb138-7"><a href="#cb138-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb138-8"><a href="#cb138-8" aria-hidden="true" tabindex="-1"></a>  marker_file_lines <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb138-9"><a href="#cb138-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (rowNum <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(marker_df)){</span>
<span id="cb138-10"><a href="#cb138-10" aria-hidden="true" tabindex="-1"></a>    marker_file_lines <span class="ot">=</span> <span class="fu">c</span>(marker_file_lines, <span class="fu">as.character</span>(<span class="fu">paste0</span>(<span class="st">"&gt;"</span>, marker_df[rowNum, <span class="st">"cell_type"</span>], <span class="at">sep =</span> <span class="st">""</span>)))</span>
<span id="cb138-11"><a href="#cb138-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb138-12"><a href="#cb138-12" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="fu">is.na</span>(marker_df[rowNum, <span class="st">"parent"</span>])) {</span>
<span id="cb138-13"><a href="#cb138-13" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb138-14"><a href="#cb138-14" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> <span class="cf">if</span>(marker_df[rowNum, <span class="st">"parent"</span>] <span class="sc">==</span> <span class="st">"root"</span>) {</span>
<span id="cb138-15"><a href="#cb138-15" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb138-16"><a href="#cb138-16" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> <span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.na</span>(marker_df[rowNum, <span class="st">"parent"</span>])) {</span>
<span id="cb138-17"><a href="#cb138-17" aria-hidden="true" tabindex="-1"></a>        marker_file_lines <span class="ot">=</span> <span class="fu">c</span>(marker_file_lines, <span class="fu">as.character</span>(<span class="fu">paste</span>(<span class="st">"subtype of:"</span>, marker_df[rowNum, <span class="st">"parent"</span>], <span class="at">sep =</span> <span class="st">" "</span>)))</span>
<span id="cb138-18"><a href="#cb138-18" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb138-19"><a href="#cb138-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb138-20"><a href="#cb138-20" aria-hidden="true" tabindex="-1"></a>    marker_file_lines <span class="ot">=</span> <span class="fu">c</span>(marker_file_lines, <span class="fu">as.character</span>(<span class="fu">paste</span>(<span class="st">"expressed:"</span>, marker_df[rowNum, <span class="st">"marker_gene"</span>], <span class="at">sep =</span> <span class="st">" "</span>)))</span>
<span id="cb138-21"><a href="#cb138-21" aria-hidden="true" tabindex="-1"></a>    marker_file_lines <span class="ot">=</span> <span class="fu">c</span>(marker_file_lines, <span class="st">""</span>)</span>
<span id="cb138-22"><a href="#cb138-22" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb138-23"><a href="#cb138-23" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb138-24"><a href="#cb138-24" aria-hidden="true" tabindex="-1"></a>  fileConn <span class="ot">&lt;-</span> <span class="fu">file</span>(filename)</span>
<span id="cb138-25"><a href="#cb138-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">writeLines</span>(marker_file_lines, fileConn)</span>
<span id="cb138-26"><a href="#cb138-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">close</span>(fileConn)</span>
<span id="cb138-27"><a href="#cb138-27" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write_garnett_marker_file</span>(marker_df_filtered_orph_rem, <span class="st">"mmBrain_markers_filtered.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="prepare-trained-classified-from-heirarchical-marker-gene-file-1" class="level3">
<h3 class="anchored" data-anchor-id="prepare-trained-classified-from-heirarchical-marker-gene-file-1">Prepare trained classified from heirarchical marker gene file</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Mm.eg.db)</span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a>marker_file_path <span class="ot">&lt;-</span> <span class="st">"mmBrain_markers_filtered.txt"</span></span>
<span id="cb140-4"><a href="#cb140-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-5"><a href="#cb140-5" aria-hidden="true" tabindex="-1"></a>marker_check <span class="ot">&lt;-</span> <span class="fu">check_markers</span>(koala_umap_cds, marker_file_path,</span>
<span id="cb140-6"><a href="#cb140-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">db =</span> org.Mm.eg.db,</span>
<span id="cb140-7"><a href="#cb140-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">cds_gene_id_type =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb140-8"><a href="#cb140-8" aria-hidden="true" tabindex="-1"></a>                              <span class="at">marker_file_gene_id_type =</span> <span class="st">"SYMBOL"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>There are 13 cell type definitions</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>marker_check_plot <span class="ot">&lt;-</span> <span class="fu">plot_markers</span>(marker_check)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The marker plot is too large to be viewed in R or in the Rmarkdown output. (It errors). Please view “mmBrain_markers_check_plot.png” in the google drive filesystem. It is exported with manually-found dimensions that work for display and do not error.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"mmBrain_markers_check_plot.png"</span>, marker_check_plot, <span class="at">width =</span> <span class="dv">10</span>, <span class="at">height =</span> <span class="dv">20</span>, <span class="at">limitsize =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Basically, this is a really big model.</p>
</section>
<section id="train-the-classifier" class="level3">
<h3 class="anchored" data-anchor-id="train-the-classifier">Train the classifier</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="co"># import previoously trained classifier</span></span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a>mmBrain_classifier <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../../pdcherry_github_data/scRNAseq/2023_06_03-13_cell_def_mmBrain_classifier.Rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="perform-garnett-classify" class="level3">
<h3 class="anchored" data-anchor-id="perform-garnett-classify">Perform garnett classify</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>koala_features_classified <span class="ot">&lt;-</span> <span class="fu">classify_cells</span>(koala_umap_cds, mmBrain_classifier,</span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">db =</span> org.Mm.eg.db,</span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">cluster_extend =</span> <span class="cn">TRUE</span>,</span>
<span id="cb145-4"><a href="#cb145-4" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">cds_gene_id_type =</span> <span class="st">"SYMBOL"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>This graph was created by an old(er) igraph version.
  Call upgrade_graph() on it to use with the current igraph version
  For now we convert it on the fly...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a>koala_features_classified</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CellDataSet (storageMode: environment)
assayData: 17426 features, 1410 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: ATGGCTGTCGATTTGA-1 GGTACAAAGTTCCTGC-1 ...
    TCGCCCATCTAATTGG-1 (1410 total)
  varLabels: orig.ident nCount_RNA ... cluster_ext_type (9 total)
  varMetadata: labelDescription
featureData
  featureNames: Xkr4 Rp1 ... AC149090.1 (17426 total)
  fvarLabels: vf_vst_counts_mean vf_vst_counts_variance ...
    gene_short_name (9 total)
  fvarMetadata: labelDescription
experimentData: use 'experimentData(object)'
Annotation:  </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(tidyverse))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pData</span>(koala_features_classified) <span class="sc">%&gt;%</span></span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(cell_type)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                          cell_type   n
1                        Astrocytes  89
2              Astroependymal cells  33
3                          CNS glia 111
4     Di- and mesencephalon neurons  35
5            Neural crest-like glia  34
6                           Neurons  83
7        Peripheral sensory neurons  57
8  Telencephalon projecting neurons 126
9                           Unknown 830
10                   Vascular cells  12</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pData</span>(koala_features_classified) <span class="sc">%&gt;%</span></span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(cluster_ext_type)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   cluster_ext_type   n
1                        Astrocytes 645
2              Astroependymal cells  16
3                          CNS glia  46
4     Di- and mesencephalon neurons  22
5            Neural crest-like glia  32
6                           Neurons  35
7        Peripheral sensory neurons  17
8  Telencephalon projecting neurons  96
9                           Unknown 490
10                   Vascular cells  11</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>koala_features_classified_df <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(</span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="fu">as.data.frame</span>(<span class="fu">pData</span>(koala_features_classified))),</span>
<span id="cb154-3"><a href="#cb154-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="fu">as.data.frame</span>(koala_features_classified<span class="sc">@</span>reducedDimS)),</span>
<span id="cb154-4"><a href="#cb154-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="st">"rowname"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>classified_umap_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(koala_features_classified_df) <span class="sc">+</span></span>
<span id="cb155-2"><a href="#cb155-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> umap_1, <span class="at">y =</span> umap_2, <span class="at">color =</span> cluster_ext_type), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb155-3"><a href="#cb155-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb155-4"><a href="#cb155-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_d</span>(<span class="at">option =</span> <span class="st">"H"</span>)</span>
<span id="cb155-5"><a href="#cb155-5" aria-hidden="true" tabindex="-1"></a>classified_umap_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-65-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a>classified_umap_drop_unknowns_plot <span class="ot">&lt;-</span></span>
<span id="cb156-2"><a href="#cb156-2" aria-hidden="true" tabindex="-1"></a>  koala_features_classified_df <span class="sc">%&gt;%</span></span>
<span id="cb156-3"><a href="#cb156-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cluster_ext_type <span class="sc">!=</span> <span class="st">"Unknown"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb156-4"><a href="#cb156-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb156-5"><a href="#cb156-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> umap_1, <span class="at">y =</span> umap_2, <span class="at">color =</span> cluster_ext_type), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb156-6"><a href="#cb156-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb156-7"><a href="#cb156-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_d</span>(<span class="at">option =</span> <span class="st">"H"</span>)</span>
<span id="cb156-8"><a href="#cb156-8" aria-hidden="true" tabindex="-1"></a>classified_umap_drop_unknowns_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-66-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>classified_umap_drop_unknowns_plot <span class="ot">&lt;-</span></span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a>  koala_features_classified_df <span class="sc">%&gt;%</span></span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cell_type <span class="sc">!=</span> <span class="st">"Unknown"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb157-4"><a href="#cb157-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb157-5"><a href="#cb157-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> umap_1, <span class="at">y =</span> umap_2, <span class="at">color =</span> cell_type), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb157-6"><a href="#cb157-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb157-7"><a href="#cb157-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_d</span>(<span class="at">option =</span> <span class="st">"H"</span>)</span>
<span id="cb157-8"><a href="#cb157-8" aria-hidden="true" tabindex="-1"></a>classified_umap_drop_unknowns_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-67-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="htp-go-anlaysis-for-all-10-clusters" class="level2">
<h2 class="anchored" data-anchor-id="htp-go-anlaysis-for-all-10-clusters">HTP GO Anlaysis for all 10 clusters</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_for_GO <span class="ot">&lt;-</span> koala_umap_markers <span class="sc">%&gt;%</span></span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span> <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">20</span>, <span class="at">order_by =</span> avg_log2FC)</span>
<span id="cb158-3"><a href="#cb158-3" aria-hidden="true" tabindex="-1"></a>koala_umap_markers_for_GO <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="st">"gene"</span>) <span class="sc">%&gt;%</span> <span class="fu">slice_tail</span>(<span class="at">n =</span> <span class="dv">167</span>) <span class="sc">%&gt;%</span> <span class="fu">dput</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>structure(list(gene = c("Plpp1", "Kdr", "Flt4", "F8", "Fgd5", 
"Fcgr2b", "Adgrl4", "Gm35696", "Uroc1", "Cxcl1", "Tedc2", "Hal", 
"Grm8", "Gldc", "Hsd17b13", "Gcnt2", "BC024386", "Itih3", "Noct", 
"Ftcd", "Osgin1", "Amdhd1", "Gls2", "Ctcflos", "Gm50136", "Gramd1c", 
"Serpina10", "Vtn", "mt-Co1", "mt-Nd4", "Gm48099", "mt-Co2", 
"mt-Nd1", "mt-Cytb", "Serpina1c", "Il31ra", "Ces1c", "Lars2", 
"mt-Nd3", "mt-Atp6", "mt-Nd2", "Serpinf2", "Cmss1", "Cdk8", "Gm15564", 
"mt-Co3", "Serpina1e", "Nr1h5", "Mamdc2", "Arvcf", "Ngf", "Ank3", 
"Gpc6", "Pde3a", "Cacna2d1", "Bmp5", "Nrxn1", "Gucy1a1", "Fendrr", 
"Abcc9", "Ntm", "Reln", "Pde1a", "Col5a1", "Carmn", "Nkain2", 
"Lhfp", "Mup17", "mt-Cytb", "Il31ra", "mt-Atp8", "mt-Nd2", "mt-Nd4", 
"mt-Co2", "Apoe", "Ppfia2", "mt-Co3", "mt-Atp6", "Pgrmc1", "Plxnc1", 
"mt-Co1", "Dab2", "Erg", "mt-Nd5", "Chchd2", "Serpina3k", "Fcgr2b", 
"Cyp2c38", "Gm13775", "Lhpp", "Slc1a2", "Rnf43", "Prodh", "Atf3", 
"Glt1d1", "Cyp3a59", "Serpina6", "Glul", "Tox", "Slc16a10", "Klb", 
"Slc25a25", "Irs1", "Tenm3", "Kank1", "Nr0b2", "Hnf4aos", "Adgre4", 
"Fam129a", "Cd5l", "Adgb", "Bank1", "E230029C05Rik", "Clec4f", 
"Slc8a1", "Entpd1", "Abr", "Fyb", "Gm26740", "Hdac9", "Madd", 
"Unc93b1", "Elmo1", "Runx1", "Ctsc", "Slc40a1", "Cadm1", "Nuak1", 
"Wnt9b", "Vwf", "Dnm3", "Mecom", "Eln", "Thsd7a", "Rspo3", "Pgm5", 
"Unc13b", "Ptpn14", "Fbn1", "Zfp521", "Jam2", "Heg1", "Pecam1", 
"Plxna4", "March3", "Myo1d", "Tbc1d4", "Bcl11b", "Atp8a2", "Gm2682", 
"Themis", "Cd226", "Itk", "Skap1", "Ms4a4b", "Grap2", "Satb1", 
"Kcnq5", "Ptpn22", "Pde7a", "Arhgap15", "Dock2", "Inpp4b", "Epsti1", 
"Itgal", "Pik3cd", "Ripor2")), row.names = c(NA, -167L), class = c("tbl_df", 
"tbl", "data.frame"))</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">FeaturePlot</span>(koala_umap, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"Ptprb"</span>, <span class="st">"Stab2"</span>, <span class="st">"Lrmda"</span>, <span class="st">"Prkg1"</span>), <span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2024_01_22-scRNA-seq_files/figure-html/unnamed-chunk-69-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<!-- ```{r} -->
<!-- (koala_umap_GOs <- topGOtable(DEgenes = c("Hal", "Grm8", "Hsd17b13", "Sds", "Gm35696", -->
<!--                                           "Ncam2", "Chrm3", "Acaca", "Gldc", "Aox3"), -->
<!--                               BGgenes = bg_gene_vector, -->
<!--                                 #koala_js@assays$RNA@var.features, -->
<!--                               ontology = "BP", geneID = "symbol", addGeneToTerms = TRUE)) -->
<!-- ``` -->


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://utteranc.es/client.js" repo="pdcherry/blogComments" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Copyright 2024, Patrick Cherry <a rel="me" href="https://fosstodon.org/@pcherry">@pcherry@fosstodon.org</a> or <a rel="me" href="https://genomic.social/@pcherry">@pcherry@genomic.social</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>