<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Patrick Cherry">
<meta name="dcterms.date" content="2024-04-05">

<title>Frameshift – DoE: Design of Experiment</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<meta name="fediverse:creator" content="@pcherry@fosstodon.org">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
<meta property="og:title" content="Frameshift - DoE: Design of Experiment">
<meta property="og:description" content="">
<meta property="og:image" content="2024_04_05-DoE_design_of_experiment/2024_04_05-DoE_architecture_imperfect_grid.jpg">
<meta property="og:site_name" content="Frameshift">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Frameshift</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Patrick-Cherry-cv.html"> 
<span class="menu-text">CV</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/pdcherry"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://fosstodon.org/@pcherry"> <i class="bi bi-mastodon" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://pixelfed.social/pcherry"> <i class="bi bi-instagram" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/p-cherry/"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">DoE: Design of Experiment</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">DoE</div>
                <div class="quarto-category">design of experiment</div>
                <div class="quarto-category">code</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Patrick Cherry </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">April 5, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction-and-motivation" id="toc-introduction-and-motivation" class="nav-link active" data-scroll-target="#introduction-and-motivation">Introduction and motivation</a>
  <ul class="collapse">
  <li><a href="#historical-roots-of-design-of-experiment" id="toc-historical-roots-of-design-of-experiment" class="nav-link" data-scroll-target="#historical-roots-of-design-of-experiment">Historical roots of design of experiment</a></li>
  </ul></li>
  <li><a href="#brief-intro-to-doe-concepts" id="toc-brief-intro-to-doe-concepts" class="nav-link" data-scroll-target="#brief-intro-to-doe-concepts">Brief intro to DoE concepts</a>
  <ul class="collapse">
  <li><a href="#aplying-doe-principle-to-this-experiment" id="toc-aplying-doe-principle-to-this-experiment" class="nav-link" data-scroll-target="#aplying-doe-principle-to-this-experiment">Aplying DoE principle to this experiment</a></li>
  </ul></li>
  <li><a href="#procedure" id="toc-procedure" class="nav-link" data-scroll-target="#procedure">Procedure</a></li>
  <li><a href="#doe-with-blocking-for-multiple-operators" id="toc-doe-with-blocking-for-multiple-operators" class="nav-link" data-scroll-target="#doe-with-blocking-for-multiple-operators">DoE with blocking for multiple operators</a>
  <ul class="collapse">
  <li><a href="#check-orthogonality-of-blocking" id="toc-check-orthogonality-of-blocking" class="nav-link" data-scroll-target="#check-orthogonality-of-blocking">Check orthogonality of blocking</a></li>
  <li><a href="#analyze-captures" id="toc-analyze-captures" class="nav-link" data-scroll-target="#analyze-captures">Analyze captures</a></li>
  </ul></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="2024_04_05-DoE_design_of_experiment/2024_04_05-DoE_architecture_imperfect_grid.jpg" class="img-fluid figure-img"></p>
<figcaption>DoE experiments can make conclusions that are greater than the sum of its parts, through thoughtful designs. Much like an architectural facade, DoEs often employ a grid of conditions that do not perfectly sample every possible combination. Image credit: seier+seier on Flickr under CC</figcaption>
</figure>
</div>
<blockquote class="blockquote">
<p>To consult the statistician after an experiment is finished is often merely to ask him to conduct a post mortem examination. He can perhaps say what the experiment died of. — Ronald A. Fisher First Session of the Indian Statistical Conference, Calcutta, 1938 <span class="citation" data-cites="OxfordEssentialQuotations">(<a href="#ref-OxfordEssentialQuotations" role="doc-biblioref">Ratcliffe 2018</a>)</span></p>
</blockquote>
<section id="introduction-and-motivation" class="level2">
<h2 class="anchored" data-anchor-id="introduction-and-motivation">Introduction and motivation</h2>
<p>In this post, I describe an experimental design for generating a genomics dataset using statistical DoE principles.</p>
<p>My motivation for designing and carrying out this experiment was that I had data showing the sensitivity (by percent recall of a set of genetic variants expressed as mRNAs) over a controlled range of concentrations (made by serial dilution). <em>However, those data were only gathered with one target enrichment panel, which was a fairly esoteric TE Panel.</em> The goal was to extend the conclusions on sensitivity and recall to other, more mainstream panels used in RNA-seq experiments. We’ll call the esoteric panel “Panel A.” I had the idea that, given the success Panel A’s bioinformatic performance, <em>it could be useful to show that many different TE panels work well for an extension of said bioinformatic performance</em>, encouraging the use of those other panels by showing equal analytical performance in a head-to-head experiment.</p>
<p>To do so, I proposed using multiplexed capture (to save on wet-lab resources and time), 100 ng of input RNA (and optionally 10 ng), and technical replicates (n = 3, or a minimum of n = 2 for dropout protection). These are parameterized in this DoE script below, and the resulting sample plan is exported to google sheets for directing the lab activities a team of scientists and research associates.</p>
<section id="historical-roots-of-design-of-experiment" class="level3">
<h3 class="anchored" data-anchor-id="historical-roots-of-design-of-experiment">Historical roots of design of experiment</h3>
<p>British statistician and geneticist Ronald A. Fisher (1890-1962) heavily influenced the modern ideals of experimental design. His influential 1922 publication on maximum likelihood laid the groundwork for now widely-taught concepts of sufficiency, information, and efficiency <span class="citation" data-cites="Fisher1922">(<a href="#ref-Fisher1922" role="doc-biblioref">Fisher 1922</a>)</span>. Fisher also pioneered and developed the methods of study randomization, and he promoted principles of <em>design of experiment</em> <span class="citation" data-cites="Edwards1974">(<a href="#ref-Edwards1974" role="doc-biblioref">Edwards 1974</a>)</span>. The radical idea (at the time) was that designs could be based on varying multiple factors simultaneously, and the downstream analysis could be supported by multi-factor analysis of variance. This progress in design and analysis allowed for the testing of many variables and factors in one experiment (when designed properly) instead of many one-by-one factor comparisons.</p>
</section>
</section>
<section id="brief-intro-to-doe-concepts" class="level2">
<h2 class="anchored" data-anchor-id="brief-intro-to-doe-concepts">Brief intro to DoE concepts</h2>
<p>Below, I provide a glossary of design of experiment (DoE) concepts and their definitions. After, I describe which apply to here, and how I incorporated that concept into the design.</p>
<ul>
<li><em>Independent variable</em>: Also called treatment factor, it is a variable under study as a potential cause, and so it is controlled at or near some target value, as informed by the hypothesis. An experiment can have more than one independent variable; it is common to analyze multiple independent variables at the same time.</li>
<li><em>Background variable</em>: Also called a <em>lurking variable</em>, it is a variable that the experimenter is not aware exists or affects the outcome of the experiment. Due to this lack of awareness, it is not controlled. In a well-planned experiment, lurking variables should balance out in effect so as not to impact the conclusion of the experiment.</li>
<li><em>Dependent variable</em>: Also called a <em>response variable</em>, it is the outcome or output measured from the experiment. An experiment can have multiple dependent variables, but it is more typical to analyze for one of these variables at a time.</li>
<li><em>Replicate</em>: additional runs of the same experiment (same treatment, same controls, etc.). Replicates help to understand the within-condition variability of outcomes, and decide if enough evidence is present to conclude the independent variable is <em>causing</em> a change in the response variable.</li>
<li><em>Confounded factors</em>: <em>Confounding</em> is a relationship of two or more factors to each other. Two or more factors are confounded when they are <em>auto-correlated</em> in the design—that is, when the levels / values of each factor co-occur throughout the experiment—to the exclusion of other combinations of levels of those factors. In this situation, it is impossible to provide evidence for which factor causes any observed change in the response variable. (Some analyses will emit an error or a warning when confounding or autocorrelation of independent variable occurs.)</li>
<li><em>Biased factor</em>: <em>Bias</em> in another relationship between experimental factors. Bias occurs when an experimenter makes a change to an independent variable at the precise time when changes in a lurking variable also occur. This is effectively confounding with a lurking variable, but because lurking variables are not always known or measured, the situation can go undetected, which can lead to erroneous conclusions. For this reason, it can be good to measure suspected lurking variables <em>even if they are not controlled</em> for retrospective analysis of bias.</li>
<li><em>Experimental error</em>: the difference between the measured response for a condition and the long-run average of that condition. In this definition, error is a statistical measure, and not literally a “mistake” or “inaccuracy,” as synonyms for error suggest. Experimental error is expected (to some degree) because background or lurking variables can subtly influence outcomes or measurements.</li>
<li><em>Randomization</em>: the random assignment of conditions to a particular sample of an experiment, which leads to the likely reduction of experimental error.</li>
<li><em>Blocking</em>: the spreading of all other variables as evenly as is practical over the variable in question. If a variable is suspected of being a lurking variable, it can be “blocked” in the experimental design, which spreads its bias (if any) as evenly as is practical across all other combinations of factors. This even spreading allows for a bias in the blocked variable to be detected with high precision in analysis.</li>
</ul>
<p>Experimental error is subdivided into <strong>bias error</strong> and <strong>random error</strong>:</p>
<ul>
<li><em>Bias error</em> tends to remain constant or change in a consistent pattern over time or with other variables (whether measured or not).</li>
<li><em>Random error</em> changes from one experiment to another unpredictably. However, it the long run, random error averages to zero and has a consistent probability distribution that allows us to draw conclusions.</li>
</ul>
<section id="aplying-doe-principle-to-this-experiment" class="level3">
<h3 class="anchored" data-anchor-id="aplying-doe-principle-to-this-experiment">Aplying DoE principle to this experiment</h3>
<p>Randomization is a good way to get random error to cancel out within an experiment. Any one randomized experiment that suffers (by chance) from a preponderance of random error will not replicate that random error in a repeat experiment, because the randomization will change, and likely distribute the error differently across samples. Thus, randomization is effective for preventing error from confounding experimental conclusion in the long term.</p>
<p>But what if we want to prevent error from affecting <em>this experiment</em>? If we have a hypothesized source of the bias in mind, then we can deploy a <em>randomized complete block</em> design, also known as using blocking on a variable.</p>
<p>If we have a suspected influence from a variable that we’re not testing as an independent variable, we can attempt to minimize the contribution of its effect (or bias) to the experimental conclusions.</p>
</section>
</section>
<section id="procedure" class="level2">
<h2 class="anchored" data-anchor-id="procedure">Procedure</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>panel_info <span class="ot">&lt;-</span> <span class="fu">tribble</span>(</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span>panel,  <span class="sc">~</span>panel_size,  <span class="sc">~</span>needed_sequencing,</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"TE Panel A"</span>, <span class="fl">3.0</span>, <span class="cn">NA</span>,</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"TE Panel B"</span>,  <span class="fl">36.8</span>, <span class="cn">NA</span>,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"TE Panel C"</span>,  <span class="fl">35.8</span>, <span class="cn">NA</span>,</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Whole Transcriptome"</span>, <span class="cn">NA</span>, <span class="cn">NA</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="doe-with-blocking-for-multiple-operators" class="level2">
<h2 class="anchored" data-anchor-id="doe-with-blocking-for-multiple-operators">DoE with blocking for multiple operators</h2>
<p>I will block for the operators carrying out library prep, because operator is a known source of variation that is not relevant to understanding the effect of TE panel, RNA input mass, or concentration on performance. (For example, my direct report is better at target enrichment than I am—she gets better fold-80 scores and off-target percentages than me routinely. But we’re here to study <em>the panels</em>, not the operators, and we need to share the work.)</p>
<p>Blocking is the non-random assignment of samples to groups to minimize differences in the sample composition between the groups such that any effect of the grouping can be determined by the model and ignored (modeled out quantitatively and precisely).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>panels <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"TE Panel A"</span>, <span class="st">"TE Panel V"</span>, <span class="st">"TE Panel C"</span>, <span class="st">"Whole Transcriptome"</span>))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>concentrations <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(.<span class="dv">027</span>, .<span class="dv">0027</span>, .<span class="dv">00027</span>, .<span class="dv">000027</span>, .<span class="dv">0000027</span>))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>mass_inputs <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">100</span>))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>vec_levels_for_variables <span class="ot">&lt;-</span> <span class="fu">c</span>(panels, concentrations, mass_inputs)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>rna_TE_sensitivity_doe <span class="ot">&lt;-</span> <span class="fu">gen.factorial</span>(<span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">2</span>), <span class="at">nVars =</span> <span class="dv">3</span>, <span class="at">factors =</span> <span class="fu">c</span>(<span class="dv">1</span>),</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">varNames =</span> <span class="fu">c</span>(<span class="st">"panel"</span>, <span class="st">"conc"</span>, <span class="st">"mass_input"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>(., ., .) </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>rna_TE_sensitivity_doe</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["panel"],"name":[1],"type":["fct"],"align":["left"]},{"label":["conc"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["mass_input"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"1","2":"-2","3":"-1"},{"1":"2","2":"-2","3":"-1"},{"1":"3","2":"-2","3":"-1"},{"1":"4","2":"-2","3":"-1"},{"1":"1","2":"-1","3":"-1"},{"1":"2","2":"-1","3":"-1"},{"1":"3","2":"-1","3":"-1"},{"1":"4","2":"-1","3":"-1"},{"1":"1","2":"0","3":"-1"},{"1":"2","2":"0","3":"-1"},{"1":"3","2":"0","3":"-1"},{"1":"4","2":"0","3":"-1"},{"1":"1","2":"1","3":"-1"},{"1":"2","2":"1","3":"-1"},{"1":"3","2":"1","3":"-1"},{"1":"4","2":"1","3":"-1"},{"1":"1","2":"2","3":"-1"},{"1":"2","2":"2","3":"-1"},{"1":"3","2":"2","3":"-1"},{"1":"4","2":"2","3":"-1"},{"1":"1","2":"-2","3":"1"},{"1":"2","2":"-2","3":"1"},{"1":"3","2":"-2","3":"1"},{"1":"4","2":"-2","3":"1"},{"1":"1","2":"-1","3":"1"},{"1":"2","2":"-1","3":"1"},{"1":"3","2":"-1","3":"1"},{"1":"4","2":"-1","3":"1"},{"1":"1","2":"0","3":"1"},{"1":"2","2":"0","3":"1"},{"1":"3","2":"0","3":"1"},{"1":"4","2":"0","3":"1"},{"1":"1","2":"1","3":"1"},{"1":"2","2":"1","3":"1"},{"1":"3","2":"1","3":"1"},{"1":"4","2":"1","3":"1"},{"1":"1","2":"2","3":"1"},{"1":"2","2":"2","3":"1"},{"1":"3","2":"2","3":"1"},{"1":"4","2":"2","3":"1"},{"1":"1","2":"-2","3":"-1"},{"1":"2","2":"-2","3":"-1"},{"1":"3","2":"-2","3":"-1"},{"1":"4","2":"-2","3":"-1"},{"1":"1","2":"-1","3":"-1"},{"1":"2","2":"-1","3":"-1"},{"1":"3","2":"-1","3":"-1"},{"1":"4","2":"-1","3":"-1"},{"1":"1","2":"0","3":"-1"},{"1":"2","2":"0","3":"-1"},{"1":"3","2":"0","3":"-1"},{"1":"4","2":"0","3":"-1"},{"1":"1","2":"1","3":"-1"},{"1":"2","2":"1","3":"-1"},{"1":"3","2":"1","3":"-1"},{"1":"4","2":"1","3":"-1"},{"1":"1","2":"2","3":"-1"},{"1":"2","2":"2","3":"-1"},{"1":"3","2":"2","3":"-1"},{"1":"4","2":"2","3":"-1"},{"1":"1","2":"-2","3":"1"},{"1":"2","2":"-2","3":"1"},{"1":"3","2":"-2","3":"1"},{"1":"4","2":"-2","3":"1"},{"1":"1","2":"-1","3":"1"},{"1":"2","2":"-1","3":"1"},{"1":"3","2":"-1","3":"1"},{"1":"4","2":"-1","3":"1"},{"1":"1","2":"0","3":"1"},{"1":"2","2":"0","3":"1"},{"1":"3","2":"0","3":"1"},{"1":"4","2":"0","3":"1"},{"1":"1","2":"1","3":"1"},{"1":"2","2":"1","3":"1"},{"1":"3","2":"1","3":"1"},{"1":"4","2":"1","3":"1"},{"1":"1","2":"2","3":"1"},{"1":"2","2":"2","3":"1"},{"1":"3","2":"2","3":"1"},{"1":"4","2":"2","3":"1"},{"1":"1","2":"-2","3":"-1"},{"1":"2","2":"-2","3":"-1"},{"1":"3","2":"-2","3":"-1"},{"1":"4","2":"-2","3":"-1"},{"1":"1","2":"-1","3":"-1"},{"1":"2","2":"-1","3":"-1"},{"1":"3","2":"-1","3":"-1"},{"1":"4","2":"-1","3":"-1"},{"1":"1","2":"0","3":"-1"},{"1":"2","2":"0","3":"-1"},{"1":"3","2":"0","3":"-1"},{"1":"4","2":"0","3":"-1"},{"1":"1","2":"1","3":"-1"},{"1":"2","2":"1","3":"-1"},{"1":"3","2":"1","3":"-1"},{"1":"4","2":"1","3":"-1"},{"1":"1","2":"2","3":"-1"},{"1":"2","2":"2","3":"-1"},{"1":"3","2":"2","3":"-1"},{"1":"4","2":"2","3":"-1"},{"1":"1","2":"-2","3":"1"},{"1":"2","2":"-2","3":"1"},{"1":"3","2":"-2","3":"1"},{"1":"4","2":"-2","3":"1"},{"1":"1","2":"-1","3":"1"},{"1":"2","2":"-1","3":"1"},{"1":"3","2":"-1","3":"1"},{"1":"4","2":"-1","3":"1"},{"1":"1","2":"0","3":"1"},{"1":"2","2":"0","3":"1"},{"1":"3","2":"0","3":"1"},{"1":"4","2":"0","3":"1"},{"1":"1","2":"1","3":"1"},{"1":"2","2":"1","3":"1"},{"1":"3","2":"1","3":"1"},{"1":"4","2":"1","3":"1"},{"1":"1","2":"2","3":"1"},{"1":"2","2":"2","3":"1"},{"1":"3","2":"2","3":"1"},{"1":"4","2":"2","3":"1"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>rna_TE_sensitivity_doe <span class="ot">&lt;-</span> AlgDesign<span class="sc">::</span><span class="fu">optBlock</span>(<span class="at">frml =</span> <span class="sc">~</span> conc <span class="sc">+</span> panel <span class="sc">+</span> mass_input,</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>                                              <span class="at">withinData =</span> rna_TE_sensitivity_doe,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>                                              <span class="at">blocksizes =</span> <span class="fu">rep</span>(<span class="dv">60</span>, <span class="dv">2</span>), <span class="at">nRepeats =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>optBlock</code> function from the <code>AlgDesign</code> package is maximizing <span class="math inline">\(|X'X|\)</span>, the determinant of the product of the inverse of the design matrix and the design matrix, a process called <em>D-optimization</em>. (The D criterion minimizes the overall variance of the estimates during analysis.) <span class="citation" data-cites="Donev1988">(<a href="#ref-Donev1988" role="doc-biblioref">Donev and Atkinson 1988</a>)</span> Further explanation is beyond the scope of this tutorial, but is available in copious detail in the documentation of the <code>AlgDesign</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>rna_TE_sensitivity_doe<span class="sc">$</span>D;</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3789291</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>rna_TE_sensitivity_doe<span class="sc">$</span>diagonality</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.871</code></pre>
</div>
</div>
<p>Diagonality is the degree to which the blocked variables are uncorrelated: a diagonality of 1.0 is perfectly uncorrelated. A value of 0.871 is moderate. We are getting values less than 1.0, because not every number of unique sample ( 2 * 4 * <del>5</del> ) factors to be processed is divisible by the number of blocking groups. We will see this effect illustrated in the “Check orthogonality of blocking” section.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>rna_TE__blocking_df <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(<span class="fu">names</span>(rna_TE_sensitivity_doe<span class="sc">$</span>Blocks),</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                        <span class="sc">~</span> rna_TE_sensitivity_doe<span class="sc">$</span>Blocks[[.x]]) <span class="sc">%&gt;%</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>(<span class="at">.id =</span> <span class="st">"operator"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">"operator"</span> <span class="ot">=</span> LETTERS[<span class="fu">as.integer</span>(operator)],</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>         <span class="st">"operator"</span> <span class="ot">=</span> <span class="fu">paste</span>(<span class="st">"Operator"</span>, operator)) <span class="sc">%&gt;%</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(panel, conc, mass_input)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Above, for each block, we bind the rows together into one data frame, converting the block names to the intended use, which is specifying which operator is working. Then arrange the data by panel name, concentration, and mass input.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>(panels_to_join <span class="ot">&lt;-</span> rna_TE__blocking_df <span class="sc">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(panel) <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">bind_cols</span>(<span class="fu">rename</span>(panel_info, <span class="st">"panel_name"</span> <span class="ot">=</span> <span class="dv">1</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["panel"],"name":[1],"type":["fct"],"align":["left"]},{"label":["panel_name"],"name":[2],"type":["chr"],"align":["left"]},{"label":["panel_size"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["needed_sequencing"],"name":[4],"type":["lgl"],"align":["right"]}],"data":[{"1":"1","2":"TE Panel A","3":"3.0","4":"NA"},{"1":"2","2":"TE Panel B","3":"36.8","4":"NA"},{"1":"3","2":"TE Panel C","3":"35.8","4":"NA"},{"1":"4","2":"Whole Transcriptome","3":"NA","4":"NA"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Above, we prepare the panel info for joining.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>(fusconcs_to_join <span class="ot">&lt;-</span> rna_TE__blocking_df <span class="sc">%&gt;%</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(conc) <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(concentrations) <span class="sc">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="st">"concentrations"</span> <span class="ot">=</span> <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["conc"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["concentrations"],"name":[2],"type":["fct"],"align":["left"]}],"data":[{"1":"-2","2":"0.027"},{"1":"-1","2":"0.0027"},{"1":"0","2":"0.00027"},{"1":"1","2":"2.7e-05"},{"1":"2","2":"2.7e-06"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>As well as preparing the concentration info for joining.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>(massinput_to_join <span class="ot">&lt;-</span> rna_TE__blocking_df <span class="sc">%&gt;%</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(mass_input) <span class="sc">%&gt;%</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(mass_inputs) <span class="sc">%&gt;%</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="st">"mass_inputs"</span> <span class="ot">=</span> <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["mass_input"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["mass_inputs"],"name":[2],"type":["fct"],"align":["left"]}],"data":[{"1":"-1","2":"10"},{"1":"1","2":"100"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>And lastly prepare the mass input levels for joining.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>rna_TE__doe_blocked <span class="ot">&lt;-</span> rna_TE__blocking_df <span class="sc">%&gt;%</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(panels_to_join, <span class="at">by =</span> <span class="st">"panel"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(fusconcs_to_join, <span class="at">by =</span> <span class="st">"conc"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(massinput_to_join, <span class="at">by =</span> <span class="st">"mass_input"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="st">"panel"</span> <span class="ot">=</span> <span class="st">"panel_name"</span>, <span class="st">"conc"</span> <span class="ot">=</span> <span class="st">"concentrations"</span>,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>         <span class="st">"mass_input"</span> <span class="ot">=</span> <span class="st">"mass_inputs"</span>, <span class="st">"operator"</span>, panel_size, needed_sequencing) <span class="sc">%&gt;%</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(panel, <span class="fu">desc</span>(conc), mass_input) <span class="sc">%&gt;%</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">"replicate_num"</span> <span class="ot">=</span> <span class="fu">row_number</span>(), <span class="at">.by =</span> <span class="fu">c</span>(panel, conc, mass_input)) <span class="sc">%&gt;%</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(replicate_num, <span class="at">.after =</span> operator) <span class="sc">%&gt;%</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="st">"LP_operator"</span> <span class="ot">=</span> <span class="st">"operator"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(panel, mass_input) <span class="sc">%&gt;%</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># below, assign samples to captures in plexity of up to 6</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># the table is sorted in order of panel and mass input, so most of the time,</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># samples will be paired with other samples with the same mass input, which is</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># best practice.</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">"capture"</span> <span class="ot">=</span> <span class="fu">ceiling</span>(<span class="fu">row_number</span>()<span class="sc">/</span><span class="dv">6</span>) ) <span class="sc">%&gt;%</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(<span class="st">"capture"</span>, <span class="at">.after =</span> replicate_num) <span class="sc">%&gt;%</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(panel, <span class="fu">desc</span>(conc), mass_input)</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(rna_TE__doe_blocked, <span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["panel"],"name":[1],"type":["chr"],"align":["left"]},{"label":["conc"],"name":[2],"type":["fct"],"align":["left"]},{"label":["mass_input"],"name":[3],"type":["fct"],"align":["left"]},{"label":["LP_operator"],"name":[4],"type":["chr"],"align":["left"]},{"label":["replicate_num"],"name":[5],"type":["int"],"align":["right"]},{"label":["capture"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["panel_size"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["needed_sequencing"],"name":[8],"type":["lgl"],"align":["right"]}],"data":[{"1":"TE Panel A","2":"0.027","3":"10","4":"Operator A","5":"1","6":"1","7":"3","8":"NA"},{"1":"TE Panel A","2":"0.027","3":"10","4":"Operator B","5":"2","6":"1","7":"3","8":"NA"},{"1":"TE Panel A","2":"0.027","3":"10","4":"Operator B","5":"3","6":"1","7":"3","8":"NA"},{"1":"TE Panel A","2":"0.027","3":"100","4":"Operator B","5":"1","6":"3","7":"3","8":"NA"},{"1":"TE Panel A","2":"0.027","3":"100","4":"Operator B","5":"2","6":"3","7":"3","8":"NA"},{"1":"TE Panel A","2":"0.027","3":"100","4":"Operator B","5":"3","6":"3","7":"3","8":"NA"},{"1":"TE Panel A","2":"0.0027","3":"10","4":"Operator A","5":"1","6":"1","7":"3","8":"NA"},{"1":"TE Panel A","2":"0.0027","3":"10","4":"Operator A","5":"2","6":"1","7":"3","8":"NA"},{"1":"TE Panel A","2":"0.0027","3":"10","4":"Operator A","5":"3","6":"1","7":"3","8":"NA"},{"1":"TE Panel A","2":"0.0027","3":"100","4":"Operator A","5":"1","6":"4","7":"3","8":"NA"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Above, we interpret the numeric levels from the DoE by joining each to the tables that provide the literal values for those experimental variables.</p>
<p>Above, we also assigned samples to captures in plexity of up to 6 the table is sorted in order of panel and mass input, so most of the time, samples will be paired with other samples with the same mass input, which is best practice.</p>
<p>Quickly spot-checking the first 10 rows, everything looks reasonable.</p>
<section id="check-orthogonality-of-blocking" class="level3">
<h3 class="anchored" data-anchor-id="check-orthogonality-of-blocking">Check orthogonality of blocking</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>rna_TE__doe_blocked <span class="sc">%&gt;%</span> <span class="fu">count</span>(LP_operator, mass_input)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["LP_operator"],"name":[1],"type":["chr"],"align":["left"]},{"label":["mass_input"],"name":[2],"type":["fct"],"align":["left"]},{"label":["n"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"Operator A","2":"10","3":"30"},{"1":"Operator A","2":"100","3":"30"},{"1":"Operator B","2":"10","3":"30"},{"1":"Operator B","2":"100","3":"30"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Above, we count how many samples are operated by each operator at each mass level. The counts are all equal, meaning the two <code>mass_input</code> factor levels have been “spread evenly” across the available operators.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>rna_TE__doe_blocked <span class="sc">%&gt;%</span> <span class="fu">count</span>(LP_operator, panel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["LP_operator"],"name":[1],"type":["chr"],"align":["left"]},{"label":["panel"],"name":[2],"type":["chr"],"align":["left"]},{"label":["n"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"Operator A","2":"TE Panel A","3":"15"},{"1":"Operator A","2":"TE Panel B","3":"15"},{"1":"Operator A","2":"TE Panel C","3":"15"},{"1":"Operator A","2":"Whole Transcriptome","3":"15"},{"1":"Operator B","2":"TE Panel A","3":"15"},{"1":"Operator B","2":"TE Panel B","3":"15"},{"1":"Operator B","2":"TE Panel C","3":"15"},{"1":"Operator B","2":"Whole Transcriptome","3":"15"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>The same is true for the levels of the <code>panel</code> factor: they have been “spread evenly” across the available operators.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>rna_TE__doe_blocked <span class="sc">%&gt;%</span> <span class="fu">count</span>(LP_operator, conc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["LP_operator"],"name":[1],"type":["chr"],"align":["left"]},{"label":["conc"],"name":[2],"type":["fct"],"align":["left"]},{"label":["n"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"Operator A","2":"2.7e-06","3":"13"},{"1":"Operator A","2":"2.7e-05","3":"12"},{"1":"Operator A","2":"0.00027","3":"9"},{"1":"Operator A","2":"0.0027","3":"14"},{"1":"Operator A","2":"0.027","3":"12"},{"1":"Operator B","2":"2.7e-06","3":"11"},{"1":"Operator B","2":"2.7e-05","3":"12"},{"1":"Operator B","2":"0.00027","3":"15"},{"1":"Operator B","2":"0.0027","3":"10"},{"1":"Operator B","2":"0.027","3":"12"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Finally, with RNA concentration, we see that not exactly every operator has the same number of samples at each level of RNA concentration being tested. This is because the number is not evenly divisible, and so some variability in the number of assigned samples will have to exist. Qualitatively, they look reasonably evenly-spread from across the two operators.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>googlesheets4<span class="sc">::</span><span class="fu">write_sheet</span>(rna_TE__doe_blocked, <span class="at">ss =</span> <span class="st">"sheet_string_goes_here"</span>,</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">sheet =</span> <span class="st">"rna_TE_sensitivity_doe"</span>);</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>file_dest_dir <span class="ot">&lt;-</span> <span class="st">"doe_design_of_experiment_with_library_prep_files"</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_create</span>(file_dest_dir);</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(rna_TE__doe_blocked, <span class="fu">paste0</span>(file_dest_dir, <span class="st">"/"</span>,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>                                      file_pref,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">"_rna_TE_sensitivity_doe.csv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Output the design to google sheets, where the scientists and research associates can view the experiment’s sample plan.</p>
</section>
<section id="analyze-captures" class="level3">
<h3 class="anchored" data-anchor-id="analyze-captures">Analyze captures</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>rna_TE__doe_blocked <span class="sc">%&gt;%</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(capture, panel, mass_input)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["capture"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["panel"],"name":[2],"type":["chr"],"align":["left"]},{"label":["mass_input"],"name":[3],"type":["fct"],"align":["left"]},{"label":["n"],"name":[4],"type":["int"],"align":["right"]}],"data":[{"1":"1","2":"TE Panel A","3":"10","4":"6"},{"1":"2","2":"TE Panel A","3":"10","4":"6"},{"1":"3","2":"TE Panel A","3":"10","4":"3"},{"1":"3","2":"TE Panel A","3":"100","4":"3"},{"1":"4","2":"TE Panel A","3":"100","4":"6"},{"1":"5","2":"TE Panel A","3":"100","4":"6"},{"1":"6","2":"TE Panel B","3":"10","4":"6"},{"1":"7","2":"TE Panel B","3":"10","4":"6"},{"1":"8","2":"TE Panel B","3":"10","4":"3"},{"1":"8","2":"TE Panel B","3":"100","4":"3"},{"1":"9","2":"TE Panel B","3":"100","4":"6"},{"1":"10","2":"TE Panel B","3":"100","4":"6"},{"1":"11","2":"TE Panel C","3":"10","4":"6"},{"1":"12","2":"TE Panel C","3":"10","4":"6"},{"1":"13","2":"TE Panel C","3":"10","4":"3"},{"1":"13","2":"TE Panel C","3":"100","4":"3"},{"1":"14","2":"TE Panel C","3":"100","4":"6"},{"1":"15","2":"TE Panel C","3":"100","4":"6"},{"1":"16","2":"Whole Transcriptome","3":"10","4":"6"},{"1":"17","2":"Whole Transcriptome","3":"10","4":"6"},{"1":"18","2":"Whole Transcriptome","3":"10","4":"3"},{"1":"18","2":"Whole Transcriptome","3":"100","4":"3"},{"1":"19","2":"Whole Transcriptome","3":"100","4":"6"},{"1":"20","2":"Whole Transcriptome","3":"100","4":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Above, we can see that, for as many samples as possible, libraries with the same total RNA input mass are captured together in the same multiplexed hybridization reaction. In some cases, we see up to three 10 ng input libraries being co-hybridized with 100 ng input libraries. Given the number of samples we want to run and the division of different TE panels, this is the best we can do.</p>
</section>
</section>
<section id="conclusion" class="level2">
<h2 class="anchored" data-anchor-id="conclusion">Conclusion</h2>
<p>Great! We have an experiment design for testing the effect of these panels against each other and against whole transcriptome sequencing. The experiment had n = 3 replicates, and it is blocked for having two operators carry out the RNA-seq library preps.</p>
<p>Importantly, when I get feedback on the design of this sample layout, I can easily show my work. Even better, if changes are needed, the entire design is programmed, and can be changed in seconds.</p>
<p>Let’s get this experiment started.</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Donev1988" class="csl-entry" role="listitem">
Donev, A. N., and A. C. Atkinson. 1988. <span>“An Adjustment Algorithm for the Construction of Exact d -Optimum Experimental Designs.”</span> <em>Technometrics</em> 30 (4): 429–33. <a href="https://doi.org/10.1080/00401706.1988.10488438">https://doi.org/10.1080/00401706.1988.10488438</a>.
</div>
<div id="ref-Edwards1974" class="csl-entry" role="listitem">
Edwards, A. W. F. 1974. <span>“The History of Likelihood.”</span> <em>International Statistical Review / Revue Internationale de Statistique</em> 42 (1): 9–15. <a href="http://www.jstor.org/stable/1402681">http://www.jstor.org/stable/1402681</a>.
</div>
<div id="ref-fedorov1972optimal" class="csl-entry" role="listitem">
Fedorov, Valerii V, and Malyutov MB. 1972. <span>“Optimal Designs in Regression Problems.”</span>
</div>
<div id="ref-Fisher1922" class="csl-entry" role="listitem">
Fisher, Ronald A. 1922. <span>“On the Mathematical Foundations of Theoretical Statistics.”</span> <em>Philosophical Transactions of the Royal Society of London. Series A, Containing Papers of a Mathematical or Physical Character</em> 222 (594-604): 309–68.
</div>
<div id="ref-OxfordEssentialQuotations" class="csl-entry" role="listitem">
Ratcliffe, Susan. 2018. <em>Oxford Essential Quotations</em>. Oxford University Press. <a href="https://doi.org/10.1093/acref/9780191866692.001.0001">https://doi.org/10.1093/acref/9780191866692.001.0001</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://utteranc.es/client.js" repo="pdcherry/blogComments" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Copyright 2024, Patrick Cherry <a rel="me" href="https://fosstodon.org/@pcherry">@pcherry@fosstodon.org</a> or <a rel="me" href="https://genomic.social/@pcherry">@pcherry@genomic.social</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>