<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Frameshift – patrick-cherry-resume</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Frameshift</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="./about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./Patrick-Cherry-resume.html" rel="" target="" aria-current="page">
 <span class="menu-text">CV</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/pdcherry" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://genomic.social/@pcherry" rel="" target=""><i class="bi bi-mastodon" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://pixelfed.social/pcherry" rel="" target=""><i class="bi bi-instagram" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/p-cherry/" rel="" target=""><i class="bi bi-linkedin" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">



<!-- Note to self: if the error shows up: -->
<!-- In grepl("^\\s*$", x) : unable to translate to a wide string -->
<!-- re-type the sptring to avoid non-ASCII characters -->
<!-- see https://stackoverflow.com/questions/76680882/unable-to-translate-to-a-wide-string -->
<section id="aside" class="level1">
<h1>Aside</h1>
<!-- ```{r} -->
<!-- if(params$pdf_mode){ -->
<!--   cat("View this CV online with links at __") -->
<!-- } else { -->
<!--   cat("[<i class='fas fa-download'></i> Download a PDF of this CV]()") -->
<!-- } -->
<!-- ``` -->
<section id="contact" class="level2">
<h2 class="anchored" data-anchor-id="contact">Contact</h2>
<ul>
<li><i class="fa fa-envelope"></i> pcherry [at] pm [dot] me</li>
<li><i class="fa fa-phone"></i> upon request</li>
<li><i class="fa fa-suitcase"></i> Senior Scientist | Genomics</li>
<li><i class="fa fa-building"></i> Twist Bioscience</li>
<li><i class="fa fa-map"></i> San Francisco, California</li>
<li><i class="fa fa-globe"></i> <a href="https://pdcherry.github.io">pdcherry.github.io</a></li>
<li><i class="fa fa-github"></i> <a href="https://github.com/pdcherry">github.com/pdcherry</a></li>
<li><i class="fa fa-linkedin"></i> <a href="https://www.linkedin.com/in/p-cherry">linkedin.com/in/p-cherry</a></li>
</ul>
<!-- Language Skills {#skills} -->
<!-- -------------------------------------------------------------------------------- -->
<!-- ```{r skills graph} -->
<!-- CV %>% print_skill_bars() -->
<!-- ``` -->
</section>
<section id="disclaimer" class="level2">
<h2 class="anchored" data-anchor-id="disclaimer">Disclaimer</h2>
<p>Last updated on 2023-12-29.</p>
<p>Data-driven résumé made in R using pagedown.</p>
</section>
</section>
<section id="main" class="level1">
<h1>Main</h1>
<section id="patrick-cherry" class="level2">
<h2 class="anchored" data-anchor-id="patrick-cherry">Patrick Cherry</h2>
<p>PhD scientist skilled in bioinformatics, biological data science, data visualization, statistical modeling, next-generation sequencing (NGS), and tool-building. I’ve coded reproducible and rigorous pipelines for high-throughput experimental designs and genomic analyses, launched best-in-class oncology reference standards, and invented new molecular methods for DNA and microbe manipulation. Originally trained in Molecular Biology, I’m interested in taking my knowledge and practice of data science / bioinformatics to the next level, especially on spatial and single cell data analysis.</p>
</section>
<section id="education" class="level2">
<h2 class="anchored" data-anchor-id="education">Education</h2>
<section id="phd" class="level3">
<h3 class="anchored" data-anchor-id="phd">PhD</h3>
<p>University of Colorado School of Medicine</p>
<p>Aurora/Denver, Colorado</p>
<p>2019 - 2013</p>
<ul>
<li>Ph.D.&nbsp;in Molecular Biology</li>
<li>Advisor: Jay Hesselberth, PhD.</li>
<li>Thesis: RNA Terminus chemistry affects the decay events that target HAC1 mRNA during the Unfolded Protein Response</li>
</ul>
</section>
<section id="ba" class="level3">
<h3 class="anchored" data-anchor-id="ba">BA</h3>
<p>Hendrix College</p>
<p>Conway, Arkansas</p>
<p>2013 - 2009</p>
<ul>
<li>Biochemistry and Molecular Biology, with Distinction</li>
<li>Advisor: Andres Caro, PhD.</li>
<li>Senior Capstone Project showing key stress response gene expression changes to oxidative stress in liver cells</li>
<li>Minor in Mathematics; PI: Lars Seme; Project: Newton’s method as a fractal chaotic dynamical system</li>
</ul>
</section>
</section>
<section id="industry-experience" class="level2">
<h2 class="anchored" data-anchor-id="industry-experience">Industry Experience</h2>
<p>I currently split my time between wet lab and computational activities. I have worked in a variety of roles ranging from HTP strain onboarding to genomics scientist. I like collaborative environments where I can learn from my peers and in turn teach others.</p>
<section id="senior-scientist" class="level3">
<h3 class="anchored" data-anchor-id="senior-scientist">Senior Scientist</h3>
<p>Twist Bioscience</p>
<p>South San Francisco, California</p>
<p>Current - 2022</p>
<ul>
<li>Tech Lead of multiple reference control NPIs, custom OEMs, and commercial releases. Includes the <em>Pan-cancer RNA Fusion Controls</em>, <em>Fragmentome Calibration Controls</em>, <em>CNV Controls</em>, <em>Pan-cancer cfDNA v2</em>, and RNA-seq.</li>
<li>Mentored a direct report from Senior Research Associate to Scientist to serving as a Tech Lead on new product introductions</li>
<li>Original research led to multiple outside presentations and applications for patent protection of product configuration and biochemical methods.</li>
<li>Custom data analysis pipeline in R and Python demonstrated proof-of-concept design and QC success of the Pan-cancer RNA Fusion Controls; designed and implemented the production approach; used public databases and feedback from alpha testers to design configuration of fusions RNAs</li>
<li>Led new technology evaluation of a new NGS platform with custom experiments and bioinformatic analyses in Python, R, and SQL to enable faster gene QC in Production. Also led ancillary experiments to speed up synthetic gene production. Coded, implemented, and distributed on company GitHub an internal package, <em><code>twistcolorpal</code></em>, that automatically adds Twist-brand colors to <code>ggplot2</code> plots and sets up database connectors to <code>SQL</code> / Snowflake for parameterized <code>dbplyr</code> querying. Regularly use R, tidyverse, Python, Polars, AWS <code>s3</code>, <code>Spark</code>, <code>PySpark</code>, and <code>Sparklyr</code>, locally and on Databricks.</li>
</ul>
</section>
<section id="scientist" class="level3">
<h3 class="anchored" data-anchor-id="scientist">Scientist</h3>
<p>Twist Bioscience</p>
<p>South San Francisco, California</p>
<p>Current - 2021</p>
<ul>
<li>Tech Lead of <em>Pan-Cancer Reference Standard</em>, an ISO-13485 synthetic positive control with 458 unique variants among 84 cancer-associated genes at six QC’d VAFs, plus a WT control, which launched in Nov of 2021</li>
<li>Designed, implemented, &amp; validated primer removal procedure for DNA standards <em>that is compatible with methylation</em></li>
<li>Devised and validated precise high-throughput DNA quantification process for accurate pooling</li>
<li>Led multiple iterations of custom NGS analysis refined the QC approach and thresholds for ensuring a contamination-free production process; extensively used data visualization to communicate complex data to cross-functional collaboration teams.</li>
<li>Made extensive use of UMI sequencing and created novel method to rigorously quantify library conversion efficiency to evaluate product and potential secondary sources</li>
</ul>
</section>
<section id="scientist-i" class="level3">
<h3 class="anchored" data-anchor-id="scientist-i">Scientist I</h3>
<p>Zymergen, Inc.</p>
<p>Emeryville, California</p>
<p>2021 - 2019</p>
<ul>
<li>Designed and implemented an automated high-throughput genotyping assay</li>
<li>Designed &amp; carried out complex experiments on automation with and without LIMS sample tracking</li>
<li>Supported a company-wide NGS core under high demand from diverse groups with complex needs using data-driven decision making and teaching</li>
<li>Used statistical methods to screen and optimize a genetic engineering protocol for newly-on-boarded microbe; delivered robust process while working on New Product Introduction team</li>
<li>Built hundreds of plasmids using modern molecular cloning techniques like Gibson and Golden Gate</li>
</ul>
</section>
</section>
<section id="research-experience" class="level2">
<h2 class="anchored" data-anchor-id="research-experience">Research Experience</h2>
<p>I worked on a few projects during my PhD, and the RNA repair project led me to custom 5´-OH RNA-seq libraries, which inspired my fascination with transcriptomics and bioinformatics.</p>
<section id="doctoral-research" class="level3">
<h3 class="anchored" data-anchor-id="doctoral-research">Doctoral Research</h3>
<p>University of Colorado School of Medicine</p>
<p>Aurora/Denver, Colorado</p>
<p>2019 - 2014</p>
<ul>
<li>Wrote, revised, &amp; published two academic papers on RNA repair &amp; yeast genetics</li>
<li>Engineered and characterized genetic bypass of essential genes in budding yeast; on-boarded CRISPR/Cas9 for efficient and precise gene knock-in</li>
<li>Expressed, purified, and used wild-type and mutant recombinant protein in <em>E. coli</em> to carry out an RNA modification enzymatic assay</li>
<li>Optimized custom RNA-seq library protocol; independently planned, executed, troubleshooted RNA modification detection</li>
<li>Routinely conducted northern blotting, targeted depletion, primer extension, splinted ligation, and other esoteric DNA and RNA experiments</li>
</ul>
</section>
</section>
<section id="intellectual-property" class="level2">
<h2 class="anchored" data-anchor-id="intellectual-property">Intellectual Property</h2>
<p>Working at Twist and Zymergen on new product research requires confidentiality, but public evidence of accomplishments often comes in patent applications. The Legal teams know me well for being a helpful expert in the process.</p>
<section id="methylation-mediated-adapter-removal-on-nucleic-acid-sequences" class="level3">
<h3 class="anchored" data-anchor-id="methylation-mediated-adapter-removal-on-nucleic-acid-sequences">Methylation-mediated adapter removal on nucleic acid sequences</h3>
<p>Twist Bioscience</p>
<p>South San Francisco, California</p>
<p>3/7/23</p>
<ul>
<li>US 63/317,466</li>
</ul>
</section>
<section id="expansion-of-cfdna-for-libraries" class="level3">
<h3 class="anchored" data-anchor-id="expansion-of-cfdna-for-libraries">Expansion of cfDNA for Libraries</h3>
<p>Twist Bioscience</p>
<p>South San Francisco, California</p>
<p>11/12/21</p>
<ul>
<li>US Prov. Pat. Ref No 823.102</li>
</ul>
</section>
<section id="libraries-for-mutational-analysis" class="level3">
<h3 class="anchored" data-anchor-id="libraries-for-mutational-analysis">Libraries for mutational analysis</h3>
<p>Twist Bioscience</p>
<p>South San Francisco, California</p>
<p>4/9/21</p>
<ul>
<li>US Prov. Pat. Ref No 823.101</li>
</ul>
</section>
<section id="method-for-counterselection-in-microorganisms" class="level3">
<h3 class="anchored" data-anchor-id="method-for-counterselection-in-microorganisms">Method for counterselection in microorganisms</h3>
<p>Zymergen, Inc.</p>
<p>Emeryville, California</p>
<p>3/25/21</p>
<ul>
<li>US 2021_0087586 A1</li>
</ul>
</section>
</section>
<section id="selected-publications" class="level2">
<h2 class="anchored" data-anchor-id="selected-publications">Selected Publications</h2>
<p>I communicate my results clearly, both in writing and in live presentations. I enjoy writing research papers, but my career has required more tech notes and app notes recently.</p>
<section id="twist-pan-cancer-synthetic-reference-materials-technical-guide" class="level3">
<h3 class="anchored" data-anchor-id="twist-pan-cancer-synthetic-reference-materials-technical-guide">Twist Pan-cancer synthetic reference materials technical guide</h3>
<p><a href="https://www.twistbioscience.com/resources/guideguideline/twist-cfdna-pan-cancer-reference-standard-technical-guide">Twist Bioscience</a></p>
<p>South San Francisco, California</p>
<p>12/21/21</p>
<ul>
<li>Patrick Cherry &amp; Mike Bocek</li>
</ul>
</section>
<section id="multiple-decay-events-target-hac1-mrna-during-splicing-to-regulate-the-unfolded-protein-response" class="level3">
<h3 class="anchored" data-anchor-id="multiple-decay-events-target-hac1-mrna-during-splicing-to-regulate-the-unfolded-protein-response">Multiple decay events target HAC1 mRNA during splicing to regulate the unfolded protein response</h3>
<p><a href="https://doi.org/10.7554/eLife.42262">eLife</a></p>
<p>N/A</p>
<p>2019</p>
<ul>
<li>Cherry, P., Peach, S., &amp; Hesselberth, J.</li>
</ul>
</section>
<section id="genetic-bypass-of-essential-rna-repair-enzymes-in-budding-yeast" class="level3">
<h3 class="anchored" data-anchor-id="genetic-bypass-of-essential-rna-repair-enzymes-in-budding-yeast">Genetic bypass of essential RNA repair enzymes in budding yeast</h3>
<p><a href="https://doi.org/10.1261/rna.061788.117">RNA</a></p>
<p>N/A</p>
<p>2018</p>
<ul>
<li>Cherry, P., White, L., York, K., &amp; Hesselberth, J.</li>
</ul>
</section>
</section>
<section id="selected-presentations-talks" class="level2">
<h2 class="anchored" data-anchor-id="selected-presentations-talks">Selected Presentations &amp; Talks</h2>
<p>I give audience-centered presentations by adapting on the fly and over time to the venue and occasion. I like to <em>transfer knowledge</em> by giving methods/best practices talks whose slides can also serve as documentation.</p>
<section id="high-sensitivity-detection-of-specific-ultra-low-frequency-somatic-mutations-for-minimal-residual-disease-mrd-monitoring" class="level3">
<h3 class="anchored" data-anchor-id="high-sensitivity-detection-of-specific-ultra-low-frequency-somatic-mutations-for-minimal-residual-disease-mrd-monitoring">High sensitivity detection of specific ultra low-frequency somatic mutations for minimal residual disease (MRD) monitoring</h3>
<p>International Society of Liquid Biopsy Annual Congress</p>
<p>Madrid, Spain</p>
<p>11/19/23</p>
</section>
<section id="twist-pan-cancer-synthetic-rna-fusion-control-for-assay-development" class="level3">
<h3 class="anchored" data-anchor-id="twist-pan-cancer-synthetic-rna-fusion-control-for-assay-development">Twist pan-cancer synthetic RNA fusion control for assay development</h3>
<p><a href="https://www.twistbioscience.com/resources/poster/twist-pan-cancer-synthetic-rna-fusion-control-assay-development">American Association for Cancer Researchers</a></p>
<p>Orlando, Florida</p>
<p>4/16/23</p>
</section>
<section id="use-of-synthetic-cnv-fragments-to-mimic-copy-number-alterations-for-ctdna-reference-standards" class="level3">
<h3 class="anchored" data-anchor-id="use-of-synthetic-cnv-fragments-to-mimic-copy-number-alterations-for-ctdna-reference-standards">Use of synthetic CNV fragments to mimic copy number alterations for ctDNA reference standards</h3>
<p><a href="https://www.twistbioscience.com/resources/poster/use-synthetic-cnv-fragments-mimic-copy-number-alterations-ctdna-reference">Advances in Genome Biology and Technology</a></p>
<p>Hollywood, Florida</p>
<p>2/7/23</p>
</section>
<section id="twist-pan-cancer-synthetic-reference-materials-for-cell-free-dna-cfdna-assay-development" class="level3">
<h3 class="anchored" data-anchor-id="twist-pan-cancer-synthetic-reference-materials-for-cell-free-dna-cfdna-assay-development">Twist pan-cancer synthetic reference materials for cell-free DNA (cfDNA) assay development</h3>
<p><a href="https://www.twistbioscience.com/resources/poster/twist-pan-cancer-synthetic-reference-materials-cell-free-dna-cfdna-assay">American Association for Cancer Researchers</a></p>
<p>New Orleans, Louisiana</p>
<p>4/12/22</p>
</section>
<section id="molecular-methods-meet-the-standards-or-how-i-learned-to-stop-worrying-and-love-uv-quantification" class="level3">
<h3 class="anchored" data-anchor-id="molecular-methods-meet-the-standards-or-how-i-learned-to-stop-worrying-and-love-uv-quantification">Molecular Methods Meet the Standards: Or how I learned to stop worrying and love UV-quantification</h3>
<p>Twist R&amp;D Meeting</p>
<p>South San Francisco, CA</p>
<p>7/13/21</p>
</section>
<section id="r-use-at-zymergen" class="level3">
<h3 class="anchored" data-anchor-id="r-use-at-zymergen">R use at Zymergen</h3>
<p>Z-Tech Talk</p>
<p>Emeryville, CA</p>
<p>6/16/20</p>
</section>
<section id="data-driven-troubleshooting-of-ngs-experiments" class="level3">
<h3 class="anchored" data-anchor-id="data-driven-troubleshooting-of-ngs-experiments">Data-driven troubleshooting of NGS experiments</h3>
<p>Data Science Talk</p>
<p>Emeryville, CA</p>
<p>4/20/20</p>
<!-- Trainees & Direct Reports {data-icon=leaf} -->
<!-- -------------------------------------------------------------------------------- -->
<!-- ::: aside -->
<!-- ```{r} -->
<!-- CV %<>% print_text_block('trainees_reports_aside') -->
<!-- ``` -->
<!-- ::: -->
<!-- ```{r} -->
<!-- CV %<>% print_section('trainees_reports') -->
<!-- ``` -->
<!-- ```{r} -->
<!-- CV %<>% print_links() -->
<!-- ``` -->


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">Copyright 2023, Patrick Cherry</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>